 # Problem Set #6
**Jason Kibozi-Yocka**

**Dataset:** https://www.kaggle.com/open-source-sports/baseball-databank/data (master.csv, batting.csv, pitching.csv, fielding.csv)

**First-Round Draft Bust:** Clint Hurdle, 1975, Kanas City Royals (1st round)
<br>**Late-Round Draft Success:** Albert Pujols, 1999, St. Louis Cardinal (13th round)

## Import Libraries

In [1]:
import pandas as pd
import altair as alt

## Load Datasets

In [2]:
import chardet

In [3]:
# create function to load datasets
def loadData(filename):
    with open(filename, 'rb') as f:
        result = chardet.detect(f.read())
        data = pd.read_csv(filename, encoding=result['encoding'])
        #data.fillna(0, inplace=True)
    return data

In [4]:
# load multiple datasets
dfmaster = loadData(r'master.csv')
dfbatting = loadData(r'batting.csv')
dfpitching = loadData(r'pitching.csv')
dffielding = loadData(r'fielding.csv')

## Observe Datasets

In [5]:
# display all columns
pd.set_option('display.max_columns', None)

### "Master" Datatset

In [6]:
# master.csv (used to find playerID based on nameFirst and nameLast)
dfmaster.head()

playerID  birthYear  birthMonth  birthDay birthCountry birthState  \
0  aardsda01     1981.0        12.0      27.0          USA         CO   
1  aaronha01     1934.0         2.0       5.0          USA         AL   
2  aaronto01     1939.0         8.0       5.0          USA         AL   
3   aasedo01     1954.0         9.0       8.0          USA         CA   
4   abadan01     1972.0         8.0      25.0          USA         FL   

    birthCity  deathYear  deathMonth  deathDay deathCountry deathState  \
0      Denver        NaN         NaN       NaN          NaN        NaN   
1      Mobile        NaN         NaN       NaN          NaN        NaN   
2      Mobile     1984.0         8.0      16.0          USA         GA   
3      Orange        NaN         NaN       NaN          NaN        NaN   
4  Palm Beach        NaN         NaN       NaN          NaN        NaN   

  deathCity nameFirst nameLast       nameGiven  weight  height bats throws  \
0       NaN     David  Aardsma     David Allan   220.0    75.0    R      R   
1       NaN      Hank    Aaron     Henry Louis   180.0    72.0    R      R   
2   Atlanta    Tommie    Aaron      Tommie Lee   190.0    75.0    R      R   
3       NaN       Don     Aase  Donald William   190.0    75.0    R      R   
4       NaN      Andy     Abad   Fausto Andres   184.0    73.0    L      L   

        debut   finalGame   retroID    bbrefID  
0  2004-04-06  2015-08-23  aardd001  aardsda01  
1  1954-04-13  1976-10-03  aaroh101  aaronha01  
2  1962-04-10  1971-09-26  aarot101  aaronto01  
3  1977-07-26  1990-10-03  aased001   aasedo01  
4  2001-09-10  2006-04-13  abada001   abadan01

### Batting Datatset

In [7]:
# battling.csv (batting stats)
dfbatting.head()

playerID  yearID  stint teamID lgID   G     AB     R     H    2B   3B  \
0  abercda01    1871      1    TRO  NaN   1    4.0   0.0   0.0   0.0  0.0   
1   addybo01    1871      1    RC1  NaN  25  118.0  30.0  32.0   6.0  0.0   
2  allisar01    1871      1    CL1  NaN  29  137.0  28.0  40.0   4.0  5.0   
3  allisdo01    1871      1    WS3  NaN  27  133.0  28.0  44.0  10.0  2.0   
4  ansonca01    1871      1    RC1  NaN  25  120.0  29.0  39.0  11.0  3.0   

    HR   RBI   SB   CS   BB   SO  IBB  HBP  SH  SF  GIDP  
0  0.0   0.0  0.0  0.0  0.0  0.0  NaN  NaN NaN NaN   NaN  
1  0.0  13.0  8.0  1.0  4.0  0.0  NaN  NaN NaN NaN   NaN  
2  0.0  19.0  3.0  1.0  2.0  5.0  NaN  NaN NaN NaN   NaN  
3  2.0  27.0  1.0  1.0  0.0  2.0  NaN  NaN NaN NaN   NaN  
4  0.0  16.0  6.0  2.0  2.0  1.0  NaN  NaN NaN NaN   NaN

### Pitching Datatset

In [8]:
# pitching.csv (pitching stats)
dfpitching.head()

playerID  yearID  stint teamID lgID   W   L   G  GS  CG  SHO  SV  IPouts  \
0  bechtge01    1871      1    PH1  NaN   1   2   3   3   2    0   0    78.0   
1  brainas01    1871      1    WS3  NaN  12  15  30  30  30    0   0   792.0   
2  fergubo01    1871      1    NY2  NaN   0   0   1   0   0    0   0     3.0   
3  fishech01    1871      1    RC1  NaN   4  16  24  24  22    1   0   639.0   
4  fleetfr01    1871      1    NY2  NaN   0   1   1   1   1    0   0    27.0   

     H   ER  HR  BB  SO  BAOpp    ERA  IBB  WP  HBP  BK  BFP  GF    R  SH  SF  \
0   43   23   0  11   1    NaN   7.96  NaN NaN  NaN   0  NaN NaN   42 NaN NaN   
1  361  132   4  37  13    NaN   4.50  NaN NaN  NaN   0  NaN NaN  292 NaN NaN   
2    8    3   0   0   0    NaN  27.00  NaN NaN  NaN   0  NaN NaN    9 NaN NaN   
3  295  103   3  31  15    NaN   4.35  NaN NaN  NaN   0  NaN NaN  257 NaN NaN   
4   20   10   0   3   0    NaN  10.00  NaN NaN  NaN   0  NaN NaN   21 NaN NaN   

   GIDP  
0   NaN  
1   NaN  
2   NaN  
3   NaN  
4   NaN

### Fielding Datatset

In [9]:
# fielding.csv (fielding stats)
dffielding.head()

playerID  yearID  stint teamID lgID POS   G  GS  InnOuts    PO     A  \
0  abercda01    1871      1    TRO  NaN  SS   1 NaN      NaN   1.0   3.0   
1   addybo01    1871      1    RC1  NaN  2B  22 NaN      NaN  67.0  72.0   
2   addybo01    1871      1    RC1  NaN  SS   3 NaN      NaN   8.0  14.0   
3  allisar01    1871      1    CL1  NaN  2B   2 NaN      NaN   1.0   4.0   
4  allisar01    1871      1    CL1  NaN  OF  29 NaN      NaN  51.0   3.0   

      E   DP  PB  WP  SB  CS  ZR  
0   2.0  0.0 NaN NaN NaN NaN NaN  
1  42.0  5.0 NaN NaN NaN NaN NaN  
2   7.0  0.0 NaN NaN NaN NaN NaN  
3   0.0  0.0 NaN NaN NaN NaN NaN  
4   7.0  1.0 NaN NaN NaN NaN NaN

## Clean Datasets

### Main Players

In [10]:
p_botch = ['Clint', 'Hurdle'] #
p_success = ['Albert', 'Pujols'] #15

In [11]:
# find rows belonging to Danny Goodwin and Mike Cameron
dfplayers = dfmaster.loc[dfmaster['nameFirst'] == p_botch[0]].loc[dfmaster['nameLast'] == p_botch[1]].append(dfmaster
                    .loc[dfmaster['nameFirst'] == p_success[0]].loc[dfmaster['nameLast'] == p_success[1]]).reset_index().drop(
    columns=['index', 'birthYear', 'birthMonth', 'birthDay', 'birthCountry', 'birthState', 'birthCity', 'deathYear',
             'deathMonth', 'deathDay', 'deathCountry', 'deathState', 'deathCity'])

In [12]:
dfplayers
# now we have the playerIDs we need

playerID nameFirst nameLast        nameGiven  weight  height bats throws  \
0  hurdlcl01     Clint   Hurdle  Clinton Merrick   195.0    75.0    L      R   
1  pujolal01    Albert   Pujols     Jose Alberto   230.0    75.0    R      R   

        debut   finalGame   retroID    bbrefID  
0  1977-09-18  1987-06-26  hurdc001  hurdlcl01  
1  2001-04-02  2015-10-04  pujoa001  pujolal01

In [13]:
p_botch = dfplayers['playerID'][0]
p_success = dfplayers['playerID'][1]

In [14]:
# create function to filter other datasets using playerID
def filterData(data, player1, player2):
    return data.loc[data['playerID'] == player1].append(data.loc[data['playerID'] == player2]).reset_index()

### Batting

In [15]:
# find rows with batting data for Danny Goodwin and Mike Cameron
dfmasterbat = dfbatting
dfbatting = filterData(dfbatting, p_botch, p_success)

In [16]:
dfbatting

index   playerID  yearID  stint teamID lgID    G     AB      R      H  \
0    54879  hurdlcl01    1977      1    KCA   AL    9   26.0    5.0    8.0   
1    55847  hurdlcl01    1978      1    KCA   AL  133  417.0   48.0  110.0   
2    56797  hurdlcl01    1979      1    KCA   AL   59  171.0   16.0   41.0   
3    57751  hurdlcl01    1980      1    KCA   AL  130  395.0   50.0  116.0   
4    58690  hurdlcl01    1981      1    KCA   AL   28   76.0   12.0   25.0   
5    59650  hurdlcl01    1982      1    CIN   NL   19   34.0    2.0    7.0   
6    60661  hurdlcl01    1983      1    NYN   NL   13   33.0    3.0    6.0   
7    62656  hurdlcl01    1985      1    NYN   NL   43   82.0    7.0   16.0   
8    63656  hurdlcl01    1986      1    SLN   NL   78  154.0   18.0   30.0   
9    64687  hurdlcl01    1987      1    NYN   NL    3    3.0    1.0    1.0   
10   81604  pujolal01    2001      1    SLN   NL  161  590.0  112.0  194.0   
11   82934  pujolal01    2002      1    SLN   NL  157  590.0  118.0  185.0   
12   84248  pujolal01    2003      1    SLN   NL  157  591.0  137.0  212.0   
13   85642  pujolal01    2004      1    SLN   NL  154  592.0  133.0  196.0   
14   86946  pujolal01    2005      1    SLN   NL  161  591.0  129.0  195.0   
15   88306  pujolal01    2006      1    SLN   NL  143  535.0  119.0  177.0   
16   89697  pujolal01    2007      1    SLN   NL  158  565.0   99.0  185.0   
17   91083  pujolal01    2008      1    SLN   NL  148  524.0  100.0  187.0   
18   92468  pujolal01    2009      1    SLN   NL  160  568.0  124.0  186.0   
19   93829  pujolal01    2010      1    SLN   NL  159  587.0  115.0  183.0   
20   95210  pujolal01    2011      1    SLN   NL  147  579.0  105.0  173.0   
21   96627  pujolal01    2012      1    LAA   AL  154  607.0   85.0  173.0   
22   98043  pujolal01    2013      1    LAA   AL   99  391.0   49.0  101.0   
23   99457  pujolal01    2014      1    LAA   AL  159  633.0   89.0  172.0   
24  100929  pujolal01    2015      1    LAA   AL  157  602.0   85.0  147.0   

      2B   3B    HR    RBI    SB   CS     BB    SO   IBB   HBP   SH   SF  GIDP  
0    0.0  0.0   2.0    7.0   0.0  0.0    2.0   7.0   0.0   0.0  0.0  0.0   0.0  
1   25.0  5.0   7.0   56.0   1.0  3.0   56.0  84.0   1.0   1.0  1.0  6.0   6.0  
2   10.0  3.0   3.0   30.0   0.0  1.0   28.0  24.0   4.0   1.0  0.0  4.0   4.0  
3   31.0  2.0  10.0   60.0   0.0  0.0   34.0  61.0   5.0   2.0  2.0  5.0  11.0  
4    3.0  1.0   4.0   15.0   0.0  0.0   13.0  10.0   3.0   0.0  0.0  0.0   2.0  
5    1.0  0.0   0.0    1.0   0.0  1.0    2.0   6.0   2.0   1.0  0.0  0.0   1.0  
6    2.0  0.0   0.0    2.0   0.0  0.0    2.0  10.0   0.0   0.0  0.0  0.0   1.0  
7    4.0  0.0   3.0    7.0   0.0  1.0   13.0  20.0   3.0   1.0  1.0  0.0   1.0  
8    5.0  1.0   3.0   15.0   0.0  0.0   26.0  38.0   0.0   1.0  1.0  2.0   2.0  
9    0.0  0.0   0.0    0.0   0.0  0.0    0.0   1.0   0.0   0.0  0.0  0.0   0.0  
10  47.0  4.0  37.0  130.0   1.0  3.0   69.0  93.0   6.0   9.0  1.0  7.0  21.0  
11  40.0  2.0  34.0  127.0   2.0  4.0   72.0  69.0  13.0   9.0  0.0  4.0  20.0  
12  51.0  1.0  43.0  124.0   5.0  1.0   79.0  65.0  12.0  10.0  0.0  5.0  13.0  
13  51.0  2.0  46.0  123.0   5.0  5.0   84.0  52.0  12.0   7.0  0.0  9.0  21.0  
14  38.0  2.0  41.0  117.0  16.0  2.0   97.0  65.0  27.0   9.0  0.0  3.0  19.0  
15  33.0  1.0  49.0  137.0   7.0  2.0   92.0  50.0  28.0   4.0  0.0  3.0  20.0  
16  38.0  1.0  32.0  103.0   2.0  6.0   99.0  58.0  22.0   7.0  0.0  8.0  27.0  
17  44.0  0.0  37.0  116.0   7.0  3.0  104.0  54.0  34.0   5.0  0.0  8.0  16.0  
18  45.0  1.0  47.0  135.0  16.0  4.0  115.0  64.0  44.0   9.0  0.0  8.0  23.0  
19  39.0  1.0  42.0  118.0  14.0  4.0  103.0  76.0  38.0   4.0  0.0  6.0  23.0  
20  29.0  0.0  37.0   99.0   9.0  1.0   61.0  58.0  15.0   4.0  0.0  7.0  29.0  
21  50.0  0.0  30.0  105.0   8.0  1.0   52.0  76.0  16.0   5.0  0.0  6.0  19.0  
22  19.0  0.0  17.0   64.0   1.0  1.0   40.0  55.0   8.0   5.0  0.0  7.0  18.0  
23  37.0  1.0  28.0  105.0   5.0

**Table Glossary**
<br>2B : Double
<br>3B : Triple
<br>AB : At-bat
<br>BB : Walk
<br>CS : Caught Stealing
<br>G : Games Played
<br>GIDP : Ground Into Double Play
<br>H : Hit
<br>HBP : Hit-by-pitch
<br>HR : Home Run
<br>IBB : Intentional Walk
<br>R : Run
<br>RBI : Runs Batted In
<br>SB : Stolen Base
<br>SF : Sacrifice Fly
<br>SH : Sacrifice Bunt
<br>SO : Strikeout

### Pitching

In [17]:
# find rows with pitching data for Danny Goodwin and Mike Cameron
dfmasterpit = dfpitching
dfpitching = filterData(dfpitching, p_botch, p_success)

In [18]:
dfpitching

Empty DataFrame
Columns: [index, playerID, yearID, stint, teamID, lgID, W, L, G, GS, CG, SHO, SV, IPouts, H, ER, HR, BB, SO, BAOpp, ERA, IBB, WP, HBP, BK, BFP, GF, R, SH, SF, GIDP]
Index: []

Notice that nothing appears, meaning that neither have pitching data (aka neither were ever pitchers).

### Fielding

In [19]:
# find rows with fielding data for Danny Goodwin and Mike Cameron
dfmasterfiel = dffielding
dffielding = filterData(dffielding, p_botch, p_success)

In [20]:
dffielding

index   playerID  yearID  stint teamID lgID POS    G    GS  InnOuts  \
0    84482  hurdlcl01    1977      1    KCA   AL  OF    9   6.0    185.0   
1    84483  hurdlcl01    1977      1    KCA   AL  RF    9   6.0    185.0   
2    86409  hurdlcl01    1978      1    KCA   AL  1B   52  49.0   1189.0   
3    86410  hurdlcl01    1978      1    KCA   AL  3B    1   0.0      3.0   
4    86411  hurdlcl01    1978      1    KCA   AL  DH    1   NaN      NaN   
..     ...        ...     ...    ...    ...  ...  ..  ...   ...      ...   
61  164852  pujolal01    2013      1    LAA   AL  1B   34   NaN    906.0   
62  167275  pujolal01    2014      1    LAA   AL  1B  116   NaN   3051.0   
63  167276  pujolal01    2014      1    LAA   AL  3B    1   NaN     18.0   
64  169009  pujolal01    2015      1    LAA   AL  1B   95   NaN      NaN   
65  169347  pujolal01    2015      1    LAA   AL  3B    1   NaN      NaN   

       PO     A    E    DP  PB  WP  SB  CS  ZR  
0    17.0   0.0  0.0   0.0 NaN NaN NaN NaN NaN  
1    16.0   0.0  0.0   0.0 NaN NaN NaN NaN NaN  
2   412.0  25.0  6.0  47.0 NaN NaN NaN NaN NaN  
3     0.0   0.0  0.0   0.0 NaN NaN NaN NaN NaN  
4     NaN   NaN  NaN   NaN NaN NaN NaN NaN NaN  
..    ...   ...  ...   ...  ..  ..  ..  ..  ..  
61  265.0  29.0  3.0  24.0 NaN NaN NaN NaN NaN  
62  879.0  80.0  3.0  78.0 NaN NaN NaN NaN NaN  
63    0.0   1.0  0.0   0.0 NaN NaN NaN NaN NaN  
64  718.0  55.0  5.0  57.0 NaN NaN NaN NaN NaN  
65    0.0   0.0  0.0   0.0 NaN NaN NaN NaN NaN  

[66 rows x 19 columns]

Notice that our data is missing a lot of points, so lets get rid of those NaNs.

In [21]:
# get rid of the columns with NaN using the InnOuts
dffielding = dffielding[dffielding['InnOuts'].isna() == False].dropna(axis=1)

In [22]:
dffielding

index   playerID  yearID  stint teamID lgID POS    G  InnOuts      PO  \
0    84482  hurdlcl01    1977      1    KCA   AL  OF    9    185.0    17.0   
1    84483  hurdlcl01    1977      1    KCA   AL  RF    9    185.0    16.0   
2    86409  hurdlcl01    1978      1    KCA   AL  1B   52   1189.0   412.0   
3    86410  hurdlcl01    1978      1    KCA   AL  3B    1      3.0     0.0   
5    86412  hurdlcl01    1978      1    KCA   AL  LF   41    843.0    59.0   
6    86413  hurdlcl01    1978      1    KCA   AL  OF   78   1753.0   132.0   
7    86414  hurdlcl01    1978      1    KCA   AL  RF   42    910.0    75.0   
8    88258  hurdlcl01    1979      1    KCA   AL  3B    1      3.0     1.0   
10   88260  hurdlcl01    1979      1    KCA   AL  LF   30    606.0    40.0   
11   88261  hurdlcl01    1979      1    KCA   AL  OF   50   1113.0    88.0   
12   88262  hurdlcl01    1979      1    KCA   AL  RF   20    507.0    47.0   
13   90183  hurdlcl01    1980      1    KCA   AL  LF    1     15.0     0.0   
14   90184  hurdlcl01    1980      1    KCA   AL  OF  126   2931.0   233.0   
15   90185  hurdlcl01    1980      1    KCA   AL  RF  125   2916.0   236.0   
16   92080  hurdlcl01    1981      1    KCA   AL  OF   28    600.0    59.0   
17   92081  hurdlcl01    1981      1    KCA   AL  RF   28    600.0    59.0   
18   93909  hurdlcl01    1982      1    CIN   NL  LF   17    255.0    16.0   
19   93910  hurdlcl01    1982      1    CIN   NL  OF   17    258.0    17.0   
20   93911  hurdlcl01    1982      1    CIN   NL  RF    1      3.0     0.0   
21   95883  hurdlcl01    1983      1    NYN   NL  3B    9    174.0     1.0   
22   95884  hurdlcl01    1983      1    NYN   NL  OF    1     27.0     0.0   
23   95885  hurdlcl01    1983      1    NYN   NL  RF    1     27.0     0.0   
24   99877  hurdlcl01    1985      1    NYN   NL   C   17    344.0    79.0   
25   99878  hurdlcl01    1985      1    NYN   NL  LF    2     54.0     1.0   
26   99879  hurdlcl01    1985      1    NYN   NL  OF   10    222.0    10.0   
27   99880  hurdlcl01    1985      1    NYN   NL  RF    8    168.0     9.0   
28  101815  hurdlcl01    1986      1    SLN   NL  1B   39    814.0   301.0   
29  101816  hurdlcl01    1986      1    SLN   NL  3B    4     53.0     2.0   
30  101817  hurdlcl01    1986      1    SLN   NL   C    5     77.0    19.0   
31  101818  hurdlcl01    1986      1    SLN   NL  LF    5     69.0     3.0   
32  101819  hurdlcl01    1986      1    SLN   NL  OF   10    156.0    12.0   
33  101820  hurdlcl01    1986      1    SLN   NL  RF    5     87.0     9.0   
34  103874  hurdlcl01    1987      1    NYN   NL  1B    1      6.0     1.0   
35  136585  pujolal01    2001      1    SLN   NL  1B   42    861.0   283.0   
36  136586  pujolal01    2001      1    SLN   NL  3B   55   1296.0    40.0   
38  136588  pujolal01    2001      1    SLN   NL  OF   78   1836.0   128.0   
40  138936  pujolal01    2002      1    SLN   NL  1B   21    432.0   140.0   
41  138937  pujolal01    2002      1    SLN   NL  3B   41    879.0    25.0   
43  138939  pujolal01    2002      1    SLN   NL  OF  118   2622.0   173.0   
45  138941  pujolal01    2002      1    SLN   NL  SS    1      6.0     0.0   
46  141227  pujolal01    2003      1    SLN   NL  1B   62   1110.0   340.0   
48  141229  pujolal01    2003      1    SLN   NL  OF  113   2712.0   198.0   
49  143622  pujolal01    2004      1    SLN   NL  1B  150   4017.0  1458.0   
50  145851  pujolal01    2005      1    SLN   NL  1B  157   4077.0  1597.0   
51  148190  pujolal01    2006      1    SLN   NL  1B  143   3732.0  1348.0   
52  150594  pujolal01    2007      1    SLN   NL  1B  154   3975.0  1325.0   
53  152946  pujolal01    2008      1    SLN   NL  1B  144   3645.0  1297.0   
54  152947  pujolal01    2008      1    SLN   NL  2B    1      9.0     1.0   
55  155255  pujolal01    2009      1    SLN   NL  1B  159   4131.0  1473.0   
56  157580  pujolal01    2010      1    SLN   NL  1B  157   4143.0  1458.0   
57  159928  pujolal01    20

**Table Glossary**
<br>A : Assist
<br>DP : Double Play
<br>E : Error
<br>G : Games Played
<br>InnOuts : Innings\*3
<br>PO : Putout

## Analyze Datasets

## BATTING AVERAGES

*In baseball, the batting average (BA) is the number of hits divided by at bats. It is usually rounded to three decimal places and read without the decimal: A player with a batting average of .300 is "batting three-hundred." If necessary to break ties, batting averages could be taken beyond the .001 measurement. In this context, .001 is considered a "point," such that a .235 batter is 5 points higher than a .230 batter.*

![](https://wikimedia.org/api/rest_v1/media/math/render/svg/b1a9a962a9a544aa68480dfacb4a5d4f63ff92c0)

Source: https://en.wikipedia.org/wiki/Batting_average_(baseball)

### FORMULA

In [23]:
# calculate batting average: hits / at bats
dfbatting['Bat_Avg'] = dfbatting['H'] / dfbatting['AB']

### VISUALIZATION: BOTCH

In [24]:
dftemp = dfbatting[dfbatting['playerID'] == p_botch].copy()
dftemp['yearID'] = range(len(dftemp['yearID']))

In [25]:
dfavgbotch = pd.DataFrame(pd.Series({'playerID': p_botch, 'Bat_Avg': dftemp.mean()['Bat_Avg']}))
dfavgbotch = dfavgbotch.T
dfavgbotch

playerID   Bat_Avg
0  hurdlcl01  0.254483

In [26]:
mygraph = alt.Chart(dftemp).mark_bar(color='red').encode(
    x = alt.X('yearID', title='Year'),
    y = alt.Y('Bat_Avg', title='Batting Average'),
    tooltip = [alt.Y('Bat_Avg', title='Batting Average')]
)

myaverage = alt.Chart(dftemp).mark_rule(color='red').encode(
    y = alt.Y('mean(Bat_Avg)', title=''),
    tooltip = [alt.Y('mean(Bat_Avg)', title='Career Batting Average')]
)

myreg = mygraph.transform_regression('yearID', 'Bat_Avg', method='poly').mark_line(color='orange')

(mygraph + myaverage + myreg).properties(title='Clint Hurdle Batting Average')

alt.LayerChart(...)

**Insights:**
<br>1. Clint's batting average, across his entire career, is 0.254.
<br>2. According to the regression (orange curve) Clint's batting average seesawed, getting better, then worse, then better.
<br>3. Clint's last year was his best year (for batting) with a batting average of 0.333.

### VISUALIZATION: SUCCESS

In [27]:
dftemp = dfbatting[dfbatting['playerID'] == p_success].copy()
dftemp['yearID'] = range(len(dftemp['yearID']))

In [28]:
dfavgsuccess = pd.DataFrame(pd.Series({'playerID': p_success, 'Bat_Avg': dftemp.mean()['Bat_Avg']}))
dfavgsuccess = dfavgsuccess.T
dfavgsuccess

playerID   Bat_Avg
0  pujolal01  0.311633

In [29]:
mygraph = alt.Chart(dftemp).mark_bar(color='green').encode(
    x = alt.X('yearID', title='Year'),
    y = alt.Y('Bat_Avg', title='Batting Average'),
    tooltip = [alt.Y('Bat_Avg', title='Batting Average')]
)

myaverage = alt.Chart(dftemp).mark_rule(color='green').encode(
    y = alt.Y('mean(Bat_Avg)', title=''),
    tooltip = [alt.Y('mean(Bat_Avg)', title='Career Batting Average')]
)

myreg = mygraph.transform_regression('yearID', 'Bat_Avg', method='poly').mark_line(color='teal')

(mygraph + myaverage + myreg).properties(title='Albert Pujols Batting Average')

alt.LayerChart(...)

**Insights:**
<br>1. Albert's batting average, across his entire career, is 0.312.
<br>2. According to the regression (teal curve) Albert's batting average improved a little at first, but has been on the decline ever since.
<br>3. Albert's best year (in terms of batting) was his 3rd year, with a batting average of 0.359.

### VISUALIZATION: OTHER PLAYERS

In [30]:
dftemp = dfmasterbat[dfmasterbat['playerID'] != p_botch].copy()
dftemp = dftemp[dftemp['playerID'] != p_success].copy()
dftemp = dftemp[dftemp['yearID'] > 1967] # scores before 1967 don't qualify for batting titles

In [31]:
dfavg = pd.DataFrame(columns=[0,1])
dfavg

Empty DataFrame
Columns: [0, 1]
Index: []

In [32]:
n = 2000 # there are too many players (altair won't be able to handle it) so let's just compare them to 2000 other players

for player in dftemp['playerID'].unique():
    if n > 0:
        dftemp2 = dftemp[dftemp['playerID'] == player].copy().sum()
        if dftemp2['AB'] != 0:
            dftemp2['Bat_Avg'] = dftemp2['H'] / dftemp2['AB']
            if dftemp2['Bat_Avg'] <= 0.4 and dftemp2['Bat_Avg'] != 0:
                # anyone above 0.400 is an anomoly and anyone with 0.0 hasn't be up  to bat enough (or at all)
                appendlist = pd.Series([player, dftemp2['Bat_Avg']])
                dfavg = dfavg.append(appendlist, ignore_index=True)
                n -= 1
                print(str(n)+'.',player)

1999. aaronha01
1998. aaronto01
1997. abernte02
1996. adairje01
1995. adamsmi01
1994. adlesda01
1993. ageeto01
1992. aguirha01
1991. akerja01
1990. alcarlu01
1989. allenbe01
1988. allendi01
1987. allenha02
1986. alleyge01
1985. allisbo01
1984. alomasa01
1983. aloufe01
1982. alouje01
1981. alouma01
1980. alvarlu01
1979. alvisma01
1978. alyeabr01
1977. amaroru01
1976. andremi01
1975. aparilu01
1974. arciajo01
1973. arrigge01
1972. asprobo01
1971. azcuejo01
1970. bahnsst01
1969. bailebo01
1968. bakerdu01
1967. baldwda01
1966. bandosa01
1965. bankser01
1964. barbest01
1963. bartobo01
1962. batemjo01
1961. beaucji01
1960. beckegl01
1959. bedelho01
1958. belanma01
1957. bellga01
1956. benchjo01
1955. bennede01
1954. berryke01
1953. bertafr01
1952. billidi01
1951. billija01
1950. blairpa01
1949. blasiwa01
1948. blassst01
1947. blefacu01
1946. bobbra01
1945. boccajo01
1944. bolinbo01
1943. bondsbo01
1942. boozejo01
1941. boschdo01
1940. bosmadi01
1939. bosweda01
1938. bosweke01
1937. boutoji01

1480. shortch02
1479. siebeno01
1478. siebeso01
1477. simmote01
1476. simpsdi01
1475. simsdu01
1474. singebi01
1473. siskto01
1472. smithch04
1471. smithre06
1470. smithwi02
1469. snyderu01
1468. solaito01
1467. spangal01
1466. sparmjo01
1465. spencji01
1464. spiezed01
1463. spraged01
1462. stahlla01
1461. stangle01
1460. stanlmi01
1459. stargwi01
1458. staubru01
1457. stephje01
1456. stephjo02
1455. stonebi01
1454. stonege02
1453. stottme01
1452. stroued01
1451. suareke01
1450. sudakbi01
1449. sullijo07
1448. suthega01
1447. suttodo01
1446. swoboro01
1445. talbofr01
1444. tartajo01
1443. tatumja01
1442. tayloca01
1441. tayloro01
1440. tayloto02
1439. thomage01
1438. thomale03
1437. tiantlu01
1436. tillmbo01
1435. tolanbo01
1434. torboje01
1433. torrehe01
1432. torrejo01
1431. torremi01
1430. tovarce01
1429. tracedi01
1428. treshto01
1427. trevibo01
1426. uhlaete01
1425. unserde01
1424. uphamjo01
1423. upshace01
1422. valdesa01
1421. valenfr01
1420. vealebo01
1419. verbajo01
1418. vers

961. powelpa01
960. raderda01
959. raglato01
958. randlle01
957. richajr01
956. richale01
955. rosarji01
954. rosedo01
953. sandsch01
952. schluja01
951. seelbch01
950. slatoji01
949. soderer01
948. speiech01
947. stelmri01
946. stennre01
945. stonest01
944. stricji01
943. swansst01
942. taverfr01
941. theobro01
940. thomade01
939. thompmi01
938. torreru01
937. willich01
936. willoji01
935. yatesal01
934. ziskri01
933. adamsmi02
932. allenro01
931. barera01
930. barnelu01
929. bellbu01
928. bibbyji01
927. blancda01
926. blankla01
925. boonebo01
924. borgmgl01
923. brohaja01
922. bumbral01
921. busbyst01
920. cabelen01
919. coggiri01
918. coletch01
917. correvi01
916. cosgrmi01
915. downsda01
914. elliora01
913. evansdw01
912. fenwibo01
911. frailke01
910. freemji01
909. gallabo01
908. gilbrro01
907. goddajo01
906. goggich01
905. goltzda01
904. gonzafe01
903. gossari01
902. grubbjo01
901. hamilda01
900. harriro01
899. harrivi01
898. hernaru02
897. hiltoda01
896. holdsfr01
895. howardo01

406. leonaje01
405. mahlemi01
404. mallosh01
403. martisi01
402. mclaujo01
401. milleed04
400. morriji01
399. moskapa01
398. mullira01
397. murraed02
396. normada01
395. norriji01
394. norwowi01
393. oberkke01
392. oliveda01
391. paceljo01
390. pagemi02
389. parrila02
388. perlosa01
387. picciro01
386. puhlte01
385. pujollu01
384. putnapa01
383. richage01
382. romered01
381. romoen01
380. rosadlu01
379. sakatle01
378. schatda01
377. sembemi01
376. sextoji01
375. shirlbo01
374. skaggda01
373. smithke01
372. sorenla01
371. sotoma01
370. staggst01
369. sykesbo01
368. thomaro02
367. toddja01
366. trammal01
365. urreajo01
364. washiro01
363. washiul01
362. whisela01
361. whitalo01
360. whitsed01
359. wilforo01
358. willsbu01
357. wolfela01
356. woodsal01
355. zdebjo01
354. zeberge01
353. alberbu01
352. anderji01
351. bakerch02
350. baldwre01
349. baumgro01
348. beamoch02
347. benedbr01
346. bentobu01
345. berenju01
344. bernadw01
343. beswiji01
342. bochybr01
341. bombama01
340. boylado01
3

In [33]:
dfavg.columns = ['playerID', 'Bat_Avg']
dfavg

playerID   Bat_Avg
0     aaronha01  0.282529
1     aaronto01  0.237473
2     abernte02  0.098361
3     adairje01  0.235382
4     adamsmi01  0.250000
...         ...       ...
1995  linarru01  0.269725
1996   lisiri01  0.312500
1997  lubrast01  0.209040
1998  maldoca01  0.253775
1999  malerji01  0.225806

[2000 rows x 2 columns]

In [34]:
dfavg2 = dfavg.append(dfavgbotch).append(dfavgsuccess).mean()
dfavg2['playerID'] = 'all'
dfavg2 = pd.DataFrame(dfavg2).T
dfavg2

Bat_Avg playerID
0  0.206411      all

In [35]:
graph1 = alt.Chart(dfavgbotch).mark_bar(color='red').encode(
    x = alt.X('playerID', title=''),
    y = alt.Y('Bat_Avg', title='Batting Average'),
)

line1 = alt.Chart(dfavgbotch).mark_rule(color='red').encode(
    y = alt.Y('Bat_Avg', title='Batting Average'),
)

graph2 = graph3 = alt.Chart(dfavgsuccess).mark_bar(color='green').encode(
    x = alt.X('playerID', title=''),
    y = alt.Y('Bat_Avg', title='Batting Average'),
)

line2 = alt.Chart(dfavgsuccess).mark_rule(color='green').encode(
    y = alt.Y('Bat_Avg', title='Batting Average'),
)

graph3 = alt.Chart(dfavg).mark_bar(color='blue').encode(
    x = alt.X('playerID', title=''),
    y = alt.Y('Bat_Avg', title='Batting Average'),
)

line3 = alt.Chart(dfavg2).mark_rule(color='blue').encode(
    y = alt.Y('Bat_Avg', title='Batting Average'),
)

(graph1 + graph2 + graph3 + line1 + line2 + line3)

alt.LayerChart(...)

**Insights:**
<br>1. Albert's batting average (green) is very high compared to the MLB average (0.206).
<br>2. Even Clint's batting average (red) is above the MLB averge, which means that despite the fact that he is considered a first draft botch, he had a decent overall career (at least as a batter).

## HOMERUNS

*In baseball, a home run (abbreviated HR) is scored when the ball is hit in such a way that the batter is able to circle the bases and reach home safely in one play without any errors being committed by the defensive team in the process. In modern baseball, the feat is typically achieved by hitting the ball over the outfield fence between the foul poles (or making contact with either foul pole) without first touching the ground, resulting in an automatic home run. There is also the "inside-the-park" home run where the batter reaches home safely while the baseball is in play on the field. A home run with a high exit velocity and good launch angle is sometimes called a "no-doubter," because it leaves no doubt that it is going to leave the park when it leaves the bat.*

*When a home run is scored, the batter is also credited with a hit and a run scored, and an RBI for each runner that scores, including himself. Likewise, the pitcher is recorded as having given up a hit and a run, with additional runs charged for each runner that scores other than the batter.*

Source: https://en.wikipedia.org/wiki/Home_run

### FORMULA

In [36]:
# calculate homeruns per season (162 games): homeruns / at bats * 162 games
dfbatting['HRG'] = dfbatting['HR'] / dfbatting['G'] * 162

In [37]:
dfbatting

index   playerID  yearID  stint teamID lgID    G     AB      R      H  \
0    54879  hurdlcl01    1977      1    KCA   AL    9   26.0    5.0    8.0   
1    55847  hurdlcl01    1978      1    KCA   AL  133  417.0   48.0  110.0   
2    56797  hurdlcl01    1979      1    KCA   AL   59  171.0   16.0   41.0   
3    57751  hurdlcl01    1980      1    KCA   AL  130  395.0   50.0  116.0   
4    58690  hurdlcl01    1981      1    KCA   AL   28   76.0   12.0   25.0   
5    59650  hurdlcl01    1982      1    CIN   NL   19   34.0    2.0    7.0   
6    60661  hurdlcl01    1983      1    NYN   NL   13   33.0    3.0    6.0   
7    62656  hurdlcl01    1985      1    NYN   NL   43   82.0    7.0   16.0   
8    63656  hurdlcl01    1986      1    SLN   NL   78  154.0   18.0   30.0   
9    64687  hurdlcl01    1987      1    NYN   NL    3    3.0    1.0    1.0   
10   81604  pujolal01    2001      1    SLN   NL  161  590.0  112.0  194.0   
11   82934  pujolal01    2002      1    SLN   NL  157  590.0  118.0  185.0   
12   84248  pujolal01    2003      1    SLN   NL  157  591.0  137.0  212.0   
13   85642  pujolal01    2004      1    SLN   NL  154  592.0  133.0  196.0   
14   86946  pujolal01    2005      1    SLN   NL  161  591.0  129.0  195.0   
15   88306  pujolal01    2006      1    SLN   NL  143  535.0  119.0  177.0   
16   89697  pujolal01    2007      1    SLN   NL  158  565.0   99.0  185.0   
17   91083  pujolal01    2008      1    SLN   NL  148  524.0  100.0  187.0   
18   92468  pujolal01    2009      1    SLN   NL  160  568.0  124.0  186.0   
19   93829  pujolal01    2010      1    SLN   NL  159  587.0  115.0  183.0   
20   95210  pujolal01    2011      1    SLN   NL  147  579.0  105.0  173.0   
21   96627  pujolal01    2012      1    LAA   AL  154  607.0   85.0  173.0   
22   98043  pujolal01    2013      1    LAA   AL   99  391.0   49.0  101.0   
23   99457  pujolal01    2014      1    LAA   AL  159  633.0   89.0  172.0   
24  100929  pujolal01    2015      1    LAA   AL  157  602.0   85.0  147.0   

      2B   3B    HR    RBI    SB   CS     BB    SO   IBB   HBP   SH   SF  \
0    0.0  0.0   2.0    7.0   0.0  0.0    2.0   7.0   0.0   0.0  0.0  0.0   
1   25.0  5.0   7.0   56.0   1.0  3.0   56.0  84.0   1.0   1.0  1.0  6.0   
2   10.0  3.0   3.0   30.0   0.0  1.0   28.0  24.0   4.0   1.0  0.0  4.0   
3   31.0  2.0  10.0   60.0   0.0  0.0   34.0  61.0   5.0   2.0  2.0  5.0   
4    3.0  1.0   4.0   15.0   0.0  0.0   13.0  10.0   3.0   0.0  0.0  0.0   
5    1.0  0.0   0.0    1.0   0.0  1.0    2.0   6.0   2.0   1.0  0.0  0.0   
6    2.0  0.0   0.0    2.0   0.0  0.0    2.0  10.0   0.0   0.0  0.0  0.0   
7    4.0  0.0   3.0    7.0   0.0  1.0   13.0  20.0   3.0   1.0  1.0  0.0   
8    5.0  1.0   3.0   15.0   0.0  0.0   26.0  38.0   0.0   1.0  1.0  2.0   
9    0.0  0.0   0.0    0.0   0.0  0.0    0.0   1.0   0.0   0.0  0.0  0.0   
10  47.0  4.0  37.0  130.0   1.0  3.0   69.0  93.0   6.0   9.0  1.0  7.0   
11  40.0  2.0  34.0  127.0   2.0  4.0   72.0  69.0  13.0   9.0  0.0  4.0   
12  51.0  1.0  43.0  124.0   5.0  1.0   79.0  65.0  12.0  10.0  0.0  5.0   
13  51.0  2.0  46.0  123.0   5.0  5.0   84.0  52.0  12.0   7.0  0.0  9.0   
14  38.0  2.0  41.0  117.0  16.0  2.0   97.0  65.0  27.0   9.0  0.0  3.0   
15  33.0  1.0  49.0  137.0   7.0  2.0   92.0  50.0  28.0   4.0  0.0  3.0   
16  38.0  1.0  32.0  103.0   2.0  6.0   99.0  58.0  22.0   7.0  0.0  8.0   
17  44.0  0.0  37.0  116.0   7.0  3.0  104.0  54.0  34.0   5.0  0.0  8.0   
18  45.0  1.0  47.0  135.0  16.0  4.0  115.0  64.0  44.0   9.0  0.0  8.0   
19  39.0  1.0  42.0  118.0  14.0  4.0  103.0  76.0  38.0   4.0  0.0  6.0   
20  29.0  0.0  37.0   99.0   9.0  1.0   61.0  58.0  15.0   4.0  0.0  7.0   
21  50.0  0.0  30.0  105.0   8.0  1.0   52.0  76.0  16.0   5.0  0.0  6.0   
22  19.0  0.0  17.0   64.0   1.0  1.0   40.0  55.0   8.0   5.0  0.0  7.0   
23  37.0  1.0  28.0  105.0   5.0  1.0   48.0  71.0  11.0   5.0  0.0  9.0   
24  22.0  0.0  40.0   95.0   5.0  3.0   50.0  72.0  10.0   6.0  0.0  3.0   


### VISUALIZATION: BOTCH

In [38]:
dftemp = dfbatting[dfbatting['playerID'] == p_botch].copy()
startY = dftemp['yearID'].min()
dftemp['yearID'] = range(len(dftemp['yearID']))

In [39]:
dfhrgbotch = pd.DataFrame(pd.Series({'playerID': p_botch,  'startYear': startY, 'HR':dftemp['HR'].sum(),
                                       'HRG': dftemp.mean()['HRG']}))
dfhrgbotch = dfhrgbotch.T
dfhrgbotch

playerID startYear  HR      HRG
0  hurdlcl01      1977  32  10.5901

In [40]:
mygraph = alt.Chart(dftemp).mark_point(color='red').encode(
    x = alt.X('yearID', title='Year'),
    y = alt.Y('HRG', title='Average Home Runs per Season'),
    tooltip = [alt.Y('HRG', title='Average Home Runs per Season')]
)

myaverage = alt.Chart(dftemp).mark_rule(color='red').encode(
    y = alt.Y('mean(HRG)', title=''),
    tooltip = [alt.Y('mean(HRG)', title='Average Home Runs per Season')]
)

myreg = mygraph.transform_regression('yearID', 'HRG', method='poly').mark_line(color='orange')

(mygraph + myaverage + myreg).properties(title='Clint Hurdle Average Home Runs per Season')

alt.LayerChart(...)

**Insights:**
<br>1. Clint's average home runs per season have followed a downward trend throughout his career.
<br>2. Clint's average home runs per season, ascross his entire career, is 11 (that's about 1 home run every 13 games).
<br>2. Clint's highest average home runs per season was at the start of his career with around 36.
<br>3. Clint's last year was one of his worst years for homeruns, with an average home run per season value of 0.

### VISUALIZATION: SUCCESS

In [41]:
dftemp = dfbatting[dfbatting['playerID'] == p_success].copy()
startY = dftemp['yearID'].min()
dftemp['yearID'] = range(len(dftemp['yearID']))

In [42]:
dfhrgsuccess = pd.DataFrame(pd.Series({'playerID': p_success, 'startYear': startY, 'HR':dftemp['HR'].sum(),
                                       'HRG': dftemp.mean()['HRG']}))
dfhrgsuccess = dfhrgsuccess.T
dfhrgsuccess

playerID startYear   HR      HRG
0  pujolal01      2001  560  39.6987

In [43]:
mygraph = alt.Chart(dftemp).mark_point(color='green').encode(
    x = alt.X('yearID', title='Year'),
    y = alt.Y('HRG', title='Average Home Runs per Season'),
    tooltip = [alt.Y('HRG', title='Average Home Runs per Season')]
)

myaverage = alt.Chart(dftemp).mark_rule(color='green').encode(
    y = alt.Y('mean(HRG)', title=''),
    tooltip = [alt.Y('mean(HRG)', title='Average Home Runs per Season')]
)

myreg = mygraph.transform_regression('yearID', 'HRG', method='poly').mark_line(color='teal')

(mygraph + myaverage + myreg).properties(title='Albert Pujols Average Home Runs per Season')

alt.LayerChart(...)

**Insights:**
<br>1. Albert's average home runs per season has seesawed throughout his career, rising then falling, and then rising again.
<br>2. Albert's average home runs per season, ascross his entire career, is 40 (that's about 1 home run every 4 games).
<br>3. Albert's highest average home runs per season was in his 5th year with around 55.
<br>4. Albert's 12th year was his worst years for home runs, with an average home run per season value of about 28.
<br>5. Albert's worst year (as far as average home runs per season is concerned) is a good year for Clint.

### VISUALIZATION: OTHER PLAYERS

In [44]:
dftemp = dfmasterbat[dfmasterbat['playerID'] != p_botch].copy()
dftemp = dftemp[dftemp['playerID'] != p_success].copy()
dftemp = dftemp[dftemp['G'] > 0]

In [45]:
dfhrg = pd.DataFrame(columns=[0,1,2,3])
dfhrg

Empty DataFrame
Columns: [0, 1, 2, 3]
Index: []

In [46]:
n = 2000 # there are too many players (altair won't be able to handle it) so let's just compare them to 2000 other players

for player, year in dftemp.groupby('playerID')['yearID'].min().items():
    if n > 0:
        dftemp2 = dftemp[dftemp['playerID'] == player].copy()
        appenSY = dftemp2['yearID'].min()
        dftemp2 = dftemp2.sum()
        if dftemp2['G'] > 0:
            dftemp2['HRG'] = dftemp2['HR'] / dftemp2['G'] * 162
            appendlist = pd.Series([player, appenSY, dftemp2['HR'], dftemp2['HRG']])
            dfhrg = dfhrg.append(appendlist, ignore_index=True)
            n -= 1
            print(str(n)+'.',player)

1999. aardsda01
1998. aaronha01
1997. aaronto01
1996. aasedo01
1995. abadan01
1994. abadfe01
1993. abadijo01
1992. abbated01
1991. abbeybe01
1990. abbeych01
1989. abbotda01
1988. abbotfr01
1987. abbotgl01
1986. abbotje01
1985. abbotji01
1984. abbotku01
1983. abbotky01
1982. abbotod01
1981. abbotpa01
1980. aberal01
1979. abercda01
1978. abercre01
1977. abernbi01
1976. abernbr01
1975. abernte01
1974. abernte02
1973. abernwo01
1972. aberscl01
1971. ablesha01
1970. abnersh01
1969. abramca01
1968. abramge01
1967. abregjo01
1966. abreubo01
1965. abreujo01
1964. abreujo02
1963. abreuju01
1962. abreuto01
1961. abreuwi01
1960. abstebi01
1959. accarje01
1958. aceveal01
1957. acevejo01
1956. aceveju01
1955. achteaj01
1954. ackerji01
1953. ackerto01
1952. ackledu01
1951. acklefr01
1950. acostcy01
1949. acosted01
1948. acostjo01
1947. acostma01
1946. acostme01
1945. acrema01
1944. adairje01
1943. adairji01
1942. adamecr01
1941. adamsac01
1940. adamsau01
1939. adamsba01
1938. adamsbe01
1937. adamsbo

1482. austihe01
1481. austije01
1480. austiji01
1479. austiji02
1478. austiri01
1477. autryal01
1476. autrych01
1475. autrych02
1474. avenbr01
1473. averiea01
1472. averiea02
1471. averyst01
1470. averyxa01
1469. avilaal01
1468. avilabo01
1467. avilalu01
1466. avilemi01
1465. avilera01
1464. avreaja01
1463. axelrdy01
1462. axforjo01
1461. ayalabe01
1460. ayalabo01
1459. ayalalu01
1458. aybarer01
1457. aybarma01
1456. aybarwi01
1455. aydelja01
1454. ayersbi01
1453. ayersdo01
1452. aylwadi01
1451. ayraubo01
1450. ayraujo01
1449. azcuejo01
1448. azocaos01
1447. babbch01
1446. babcobo01
1445. babelo01
1444. babicjo01
1443. babinch01
1442. babitsh01
1441. backebr01
1440. backmle01
1439. backmwa01
1438. baconed01
1437. bacsimi01
1436. bacsimi02
1435. baczefr01
1434. badenbu01
1433. baderar01
1432. baderlo01
1431. badgrre01
1430. baeched01
1429. baekch01
1428. baergca01
1427. baezbe01
1426. baezda01
1425. baezja01
1424. baezjo01
1423. baezke01
1422. baezpe01
1421. bagbyji01
1420. bagbyji02
14

959. beirnke01
958. bejmaol01
957. belanma01
956. belarwa01
955. belchke01
954. belchti01
953. beldeir01
952. belfimi01
951. belinbo01
950. belinst01
949. belisma01
948. belisro01
947. belitto01
946. belivje01
945. belkti01
944. bellaan01
943. bellast01
942. bellaze01
941. bellbe01
940. bellbi01
939. bellbu01
938. bellch01
937. bellda01
936. bellde01
935. belleal01
934. beller01
933. bellfe01
932. bellfr01
931. bellga01
930. bellge01
929. bellge02
928. bellgu01
927. bellhe01
926. bellhi01
925. bellhma01
924. bellicl01
923. bellira01
922. belliro01
921. bellja01
920. bellje01
919. belljo01
918. bellju01
917. bellke01
916. bellle01
915. bellmi01
914. bellmi02
913. bellmja01
912. belloed01
911. belloro01
910. bellra01
909. bellro01
908. bellru01
907. bellte01
906. belltr01
905. belnovi01
904. belowdu01
903. beltbr01
902. beltrad01
901. beltrca01
900. beltren01
899. beltres01
898. beltrfr01
897. beltrom01
896. beltrri01
895. bemisha01
894. benarma01
893. benavfr01
892. benchjo01
891. bende

401. bostoly01
400. boswean01
399. bosweda01
398. bosweke01
397. botelde01
396. bottajo01
395. bottari01
394. botteke01
393. bottira01
392. bottoji01
391. bottsja01
390. botzbo01
389. bouchal01
388. bouchde01
387. bouched01
386. bouchme01
385. boudrlo01
384. bouldca01
383. boultja01
382. bourgja01
381. bourgst01
380. bourjch01
379. bourjpe01
378. bourju01
377. bournmi01
376. bournra01
375. bourqpa01
374. boutoji01
373. boveemi01
372. bowala01
371. bowcobe01
370. bowdemi01
369. bowdeti01
368. bowench01
367. bowency01
366. bowenro01
365. bowenry01
364. bowensa01
363. bowensa02
362. bowerbi01
361. bowerbr01
360. bowerce01
359. bowerfr01
358. bowersh01
357. bowerst01
356. bowesfr01
355. bowieji01
354. bowiemi01
353. bowkejo01
352. bowlebr01
351. bowlech01
350. bowleem01
349. bowlegr01
348. bowlist01
347. bowliwe01
346. bowmaab01
345. bowmabi01
344. bowmabo01
343. bowmabo02
342. bowmael01
341. bowmaer01
340. bowmajo01
339. bowmaro01
338. bowmasu01
337. bowsere01
336. bowsfte01
335. bowyetr0

In [47]:
dfhrg.columns = ['playerID', 'startYear', 'HR', 'HRG']
dfhrg

playerID startYear     HR        HRG
0     aardsda01      2004    0.0   0.000000
1     aaronha01      1954  755.0  37.086113
2     aaronto01      1962   13.0   4.819222
3      aasedo01      1977    0.0   0.000000
4      abadan01      2001    0.0   0.000000
...         ...       ...    ...        ...
1995  brownma01      1935    0.0   0.000000
1996  brownma02      1984    0.0   0.000000
1997  brownma03      1988    0.0   0.000000
1998  brownma04      1998    0.0   0.000000
1999  brownma05      2007    0.0   0.000000

[2000 rows x 4 columns]

In [48]:
dfhrg2 = dfhrg.append(dfhrgbotch).append(dfhrgsuccess)
dfhrg3 = pd.DataFrame(data=[[0,0,0]], columns=['playerID','HR', 'HRG'])
dfhrg3['playerID'] = 'all'
dfhrg3['HR'] = dfhrg2['HR'].sum()
dfhrg3['HRG'] = dfhrg2['HRG'].mean()
dfhrg3

playerID       HR       HRG
0      all  32377.0  3.363888

In [49]:
graph1 = alt.Chart(dfhrgbotch).mark_bar(color='red').encode(
    x = alt.X('startYear', title='Career Start'),
    y = alt.Y('HRG', title='Average Homeruns per Season'),
    tooltip = [alt.Y('mean(HRG)', title='Average Home Runs per Season')]
)

line1 = alt.Chart(dfhrgbotch).mark_rule(color='red').encode(
    y = alt.Y('HRG', title='Average Homeruns per Season'),
    tooltip = [alt.Y('mean(HRG)', title='Average Home Runs per Season')]
)

graph2 = alt.Chart(dfhrgsuccess).mark_bar(color='green').encode(
    x = alt.X('startYear', title='Career Start'),
    y = alt.Y('HRG', title='Average Homeruns per Season'),
    tooltip = [alt.Y('mean(HRG)', title='Average Home Runs per Season')]
)

line2 = alt.Chart(dfhrgsuccess).mark_rule(color='green').encode(
    y = alt.Y('HRG', title='Average Homeruns per Season'),
    tooltip = [alt.Y('mean(HRG)', title='Average Home Runs per Season')]
)

graph3 = alt.Chart(dfhrg).mark_circle(color='blue').encode(
    x = alt.X('startYear', title='Career Start', scale=alt.Scale(domain=[1860, 2020])),
    y = alt.Y('HRG', title='Average Homeruns per Season', scale=alt.Scale(domain=[0, 50])),
    tooltip = [alt.Y('mean(HRG)', title='Average Home Runs per Season')],
    size = 'HR'
)

line3 = alt.Chart(dfhrg3).mark_rule(color='blue').encode(
    y = alt.Y('HRG', title='Average Homeruns per Season'),
    tooltip = [alt.Y('mean(HRG)', title='Average Home Runs per Season')]
)

myreg = graph3.transform_regression('startYear', 'HRG', method='poly').mark_line(color='teal')

(graph1 + graph2 + graph3 + line1 + line2 + line3)

alt.LayerChart(...)

**Insights:**
<br>1. Average home runs per season for MLB is 3 (that's 1 home run every 54 games). This is probably because lots of guys have never hit home runs -- or very few home runs -- in their entire careers.
Albert's homeruns per season has seesawed thtoughout his career, rising then falling, and then rising again.
<br>2. Albert's average home runs is far above the MLB average.
<br>3. Even Clint's average home runs is ahead of the MLB average.

In [50]:
(graph3 + myreg)

alt.LayerChart(...)

**Insights:**
<br>A. Average home runs per season, in MLB, have been on the rise since the 1800s. So most of the homerun data is skewed towards later years. This could explain why the average home runs per season is low. Maybe the past is dragging the values down.

In [51]:
# 1967 is when MLB recording experienced a rise in scuritiny and best practices
graph4 = alt.Chart(dfhrg[dfhrg['startYear'] >= 1967].copy()).mark_circle(color='blue').encode(
    x = alt.X('startYear', title='Career Start', scale=alt.Scale(domain=[1960, 2020])),
    y = alt.Y('HRG', title='', scale=alt.Scale(domain=[0, 50])),
    tooltip = [alt.Y('mean(HRG)', title='Average Home Runs per Season')],
    size = 'HR'
)

line4 = alt.Chart(dfhrg[dfhrg['startYear'] >= 1967].copy()).mark_rule(color='blue').encode(
    y=alt.Y('mean(HRG)', title=''),
    tooltip = [alt.Y('mean(HRG)', title='Average Home Runs per Season')]
)

myreg = graph4.transform_regression('startYear', 'HRG', method='poly').mark_line(color='teal')

(graph4 + line4 + line1 + line2)

alt.LayerChart(...)

**Insights:**
<br>B. Maybe not, since limiting the start period to 1967 still only produces an average home runs per season of 4.

## RANGE FACTORS

*Range Factor (commonly abbreviated RF) is a baseball statistic developed by Bill James. It is calculated by dividing putouts and assists by the number of innings or games played at a given defense position. The statistic is premised on the notion that the total number of outs in which a player participates is more relevant in evaluating that player's defensive play than the percentage of cleanly handled chances as calculated by the conventional statistic fielding percentage.*

Source: https://en.wikipedia.org/wiki/Range_factor

### FORMULA

In [52]:
# calculate fieldig average: (putouts + assists) / games played at position
dffielding['RF'] = sum([dffielding['PO'], dffielding['A']]) / dffielding['G']

### VISUALIZATION: BOTCH

In [53]:
dftemp = dffielding[dffielding['playerID'] == p_botch].copy()
dftemp = dftemp[dftemp['RF'].notna()]
dftemp['yearID'] = range(len(dftemp['yearID']))

In [54]:
dffactorbotch = pd.DataFrame(pd.Series({'playerID': p_botch, 'G': dftemp.sum()['G'], 'RF': dftemp.mean()['RF']}))
dffactorbotch = dffactorbotch.T
dffactorbotch

playerID    G       RF
0  hurdlcl01  793  1.92324

In [55]:
mygraph = alt.Chart(dftemp).mark_line(point=True).encode(
    x = alt.X('yearID', title='Year'),
    y = alt.Y('RF', title='Range Factor'),
    color = 'POS',
    tooltip = [alt.Y('RF', title='Range Factor')]
)

myaverage = alt.Chart(dftemp).mark_rule().encode(
    y = alt.Y('mean(RF)', title=''),
    color = 'POS',
    tooltip = [alt.Y('mean(RF)', title='Career Range Factor')]
)

(mygraph + myaverage).properties(title='Clint Hurdle Range Factors')

alt.LayerChart(...)

**Insights:**
<br>1. Clints's overall range factor, taking into account all positions he's ever played, is 1.923.
<br>2. Clint's best overall performance was when he was on first base (range factor: 5.895) and catcher (range factor: 4.500).
<br>2. Clint's worst overall performance was when he was on left field (range factor: 0.930).

### VISUALIZATION: SUCCESS

In [56]:
dftemp = dffielding[dffielding['playerID'] == p_success].copy()
dftemp = dftemp[dftemp['RF'].notna()]
dftemp['yearID'] = range(len(dftemp['yearID']))

In [57]:
dffactorsuccess = pd.DataFrame(pd.Series({'playerID': p_success, 'G': dftemp.sum()['G'],'RF': dftemp.mean()['RF']}))
dffactorsuccess = dffactorsuccess.T
dffactorsuccess

playerID     G       RF
0  pujolal01  2023  5.96832

In [58]:
mygraph = alt.Chart(dftemp).mark_line(point=True).encode(
    x = alt.X('yearID', title='Year'),
    y = alt.Y('RF', title='Range Factor'),
    color = 'POS',
    tooltip = [alt.Y('RF', title='Range Factor')]
)

myaverage = alt.Chart(dftemp).mark_rule().encode(
    y = alt.Y('mean(RF)', title=''),
    color = 'POS',
    tooltip = [alt.Y('mean(RF)', title='Career Range Factor')]
)

(mygraph + myaverage).properties(title='Albert Pujols Range Factors')

alt.LayerChart(...)

**Insights:**
<br>1. Albert's overall range factor, taking into account all positions he's ever played, is 5.968.
<br>2. Albert's best overall performance was when he was on first base (range factor: 9.136).
<br>2. Albert's worst overall performance was when he was on second base (range factor: 1.000).

### VISUALIZATION: OTHERS

In [59]:
dftemp = dfmasterfiel[dfmasterfiel['playerID'] != p_botch].copy()
dftemp = dftemp[dftemp['playerID'] != p_success].copy()

In [60]:
dftemp

playerID  yearID  stint teamID lgID POS   G  GS  InnOuts    PO     A  \
0       abercda01    1871      1    TRO  NaN  SS   1 NaN      NaN   1.0   3.0   
1        addybo01    1871      1    RC1  NaN  2B  22 NaN      NaN  67.0  72.0   
2        addybo01    1871      1    RC1  NaN  SS   3 NaN      NaN   8.0  14.0   
3       allisar01    1871      1    CL1  NaN  2B   2 NaN      NaN   1.0   4.0   
4       allisar01    1871      1    CL1  NaN  OF  29 NaN      NaN  51.0   3.0   
...           ...     ...    ...    ...  ...  ..  ..  ..      ...   ...   ...   
170521  younger03    2015      1    ATL   NL  OF  22 NaN      NaN  45.0   0.0   
170522  younger03    2015      2    NYN   NL  OF   7 NaN      NaN   5.0   0.0   
170523  zimmery01    2015      1    WAS   NL  OF   1 NaN      NaN   0.0   0.0   
170524  zobribe01    2015      1    OAK   AL  OF  29 NaN      NaN  39.0   0.0   
170525  zobribe01    2015      2    KCA   AL  OF  20 NaN      NaN  36.0   1.0   

           E   DP  PB  WP  SB  CS  ZR  
0        2.0  0.0 NaN NaN NaN NaN NaN  
1       42.0  5.0 NaN NaN NaN NaN NaN  
2        7.0  0.0 NaN NaN NaN NaN NaN  
3        0.0  0.0 NaN NaN NaN NaN NaN  
4        7.0  1.0 NaN NaN NaN NaN NaN  
...      ...  ...  ..  ..  ..  ..  ..  
170521   0.0  0.0 NaN NaN NaN NaN NaN  
170522   0.0  0.0 NaN NaN NaN NaN NaN  
170523   0.0  0.0 NaN NaN NaN NaN NaN  
170524   3.0  0.0 NaN NaN NaN NaN NaN  
170525   1.0  0.0 NaN NaN NaN NaN NaN  

[170460 rows x 18 columns]

In [61]:
dffactor = pd.DataFrame(columns=[0,1,2])
dffactor

Empty DataFrame
Columns: [0, 1, 2]
Index: []

In [62]:
n = 2000 # there are too many players (altair won't be able to handle it) so let's just compare them to 2000 other players

for player in dftemp['playerID'].unique():
    if n > 0:
        dftemp2 = dftemp[dftemp['playerID'] == player].copy()
        dftemp2['RF'] = sum([dftemp2['PO'], dftemp2['A']]) / dftemp2['G']
        appendlist = pd.Series([player,  dftemp2['G'].sum(), dftemp2['RF'].mean()])
        dffactor = dffactor.append(appendlist, ignore_index=True)
        n -= 1
        print(str(n)+'.',player)

1999. abercda01
1998. addybo01
1997. allisar01
1996. allisdo01
1995. ansonca01
1994. armstbo01
1993. barkeal01
1992. barnero01
1991. barrebi01
1990. barrofr01
1989. bassjo01
1988. battijo01
1987. bealsto01
1986. beaveed01
1985. bechtge01
1984. bellast01
1983. berkena01
1982. berryto01
1981. berthha01
1980. biermch01
1979. birdge01
1978. birdsda01
1977. brainas01
1976. brannmi01
1975. burrohe01
1974. careyto01
1973. carleji01
1972. conefr01
1971. connone01
1970. cravebi01
1969. cuthbne01
1968. deaneha01
1967. donnepe01
1966. duffyed01
1965. eggleda01
1964. ewellge01
1963. fergubo01
1962. fishech01
1961. fislewe01
1960. fleetfr01
1959. flowedi01
1958. flynncl01
1957. foleyto01
1956. foranji01
1955. forceda01
1954. fulmech01
1953. glennjo01
1952. goldswa01
1951. gouldch01
1950. hallge01
1949. halliji01
1948. hamra01
1947. hastisc01
1946. hatfijo01
1945. heubege01
1944. highadi01
1943. hodesch01
1942. jackssa01
1941. johnsca01
1940. kellybi01
1939. kimbage01
1938. kingma01
1937. kingst01
1

1478. clarkjo01
1477. clinemo01
1476. comisch01
1475. critcmo01
1474. crottjo01
1473. dailyhu01
1472. dorrbe01
1471. doylejo02
1470. dwyerjo01
1469. dylerjo01
1468. eastha01
1467. eganji01
1466. eversto01
1465. farrebi01
1464. farrejo01
1463. forstto01
1462. fusseed01
1461. geisbi01
1460. gleasbi01
1459. greenbi01
1458. halbred01
1457. halpiji01
1456. hayesja01
1455. heckegu01
1454. hoganed01
1453. housech02
1452. irwinjo01
1451. jacobha01
1450. jonesbi01
1449. kienzbi01
1448. kinziwa01
1447. landido01
1446. lanech01
1445. mannfr01
1444. manniti01
1443. mansejo01
1442. maskrha01
1441. maskrle01
1440. mccafha01
1439. mccluha01
1438. mcginju01
1437. mclaufr01
1436. mcphebi01
1435. merried01
1434. morgabi02
1433. morrito01
1432. mortoch01
1431. muldomi01
1430. navasa01
1429. obrieja01
1428. obrieto01
1427. pfefffr01
1426. piercgr01
1425. reccijo01
1424. recciph01
1423. reynoch01
1422. robinya01
1421. rosemch01
1420. rowened01
1419. russjo01
1418. sayji01
1417. schapjo01
1416. scharni01
14

954. collvbi01
953. conwape01
952. crossam01
951. dailyco01
950. dailyed01
949. derbyge02
948. driscde02
947. drissmi01
946. fogarjo01
945. foleyjo01
944. greered01
943. hallsch01
942. hatfigi01
941. healyjo01
940. hilanjo01
939. hoffojo01
938. jones03
937. kellona01
936. kennete01
935. knoufed01
934. lovetto01
933. lyonsde01
932. mackre01
931. mappege01
930. maysal01
929. mccaf01
928. mcdersa01
927. mctamji01
926. mcveyge01
925. palme01
924. pechige01
923. piersdi01
922. ramseto01
921. reilljo02
920. rittech01
919. ryanji01
918. sewared01
917. stapljo01
916. stellbi01
915. stemmbi01
914. tatepo01
913. tenerjo01
912. thompsa01
911. visnejo01
910. wardjo03
909. wetzesh01
908. woodpe01
907. beglege01
906. bickhda01
905. bierblo01
904. bishobi01
903. blighne01
902. boyleja01
901. brookha01
900. chambel01
899. clarkbo01
898. clarked01
897. collihu01
896. connepe01
895. connore01
894. conwadi01
893. devliji02
892. donahji01
891. dugdada01
890. finlebi01
889. flynnjo02
888. fosteel01
887. fu

399. fitzgwa01
398. flynnmi01
397. foxpa01
396. freembu01
395. fulleha01
394. gormljo01
393. graulle01
392. griffcl01
391. hartto01
390. hugheji01
389. inksbe01
388. jennihu01
387. keenaki01
386. kellejo01
385. killefr01
384. killehe01
383. klingbi01
382. lakefr01
381. lallyda01
380. leipeja01
379. letchto01
378. lohmape01
377. longji01
376. lystobi01
375. maceji01
374. matthbo01
373. mcgrajo01
372. mcquama01
371. mcswepa01
370. meekijo01
369. merribi01
368. morelha01
367. murphla01
366. neweljo01
365. obriejo02
364. oconnjo01
363. pettepa01
362. plockwa01
361. quarlbi01
360. quinlfr01
359. reedeni01
358. rettgge01
357. ricksjo01
356. rooksge01
355. sayloph01
354. schuljo01
353. schuljo02
352. shearjo01
351. slagljo01
350. spurned01
349. sulliji01
348. tayloja01
347. walshjo01
346. wentzja01
345. ziesbi01
344. abbeybe01
343. birdfr01
342. campki01
341. cample01
340. cargobo01
339. clausfr01
338. cobbge01
337. colli01
336. connoje01
335. connoji01
334. crossmo01
333. cuppyni01
332. daly

In [63]:
dffactor.columns = ['playerID', 'G', 'RF']
dffactor

playerID     G        RF
0     abercda01     1  4.000000
1      addybo01   276  3.640755
2     allisar01   174  4.753143
3     allisdo01   344  3.527716
4     ansonca01  2592  5.311641
...         ...   ...       ...
1995  evansro01    84  1.686594
1996  fifieja01    69  1.751852
1997    foxbi01    47  5.686047
1998  gettmja01   196  5.018145
1999  grimejo01     3  4.000000

[2000 rows x 3 columns]

In [64]:
dffactor2 = dffactor.append(dffactorbotch).append(dffactorsuccess)
dffactor3 = pd.DataFrame(data=[[0,0,0]], columns=['playerID','G','RF'])
#dffactor2 = pd.DataFrame(dffactor2).T
dffactor3['playerID'] = 'all'
dffactor3['G'] = dffactor2['G'].sum()
dffactor3['RF'] = dffactor2['RF'].mean()
dffactor3

playerID       G       RF
0      all  429202  3.08318

In [65]:
graph1 = graph3 = alt.Chart(dffactorbotch).mark_bar(color='red').encode(
    x = alt.X('playerID', title=''),
    y = alt.Y('RF', title='Range Factor')
)

line1 = alt.Chart(dffactorbotch).mark_rule(color='red').encode(
    y = alt.Y('RF', title='Range Factor'),
)

graph2 = graph3 = alt.Chart(dffactorsuccess).mark_bar(color='green').encode(
    x = alt.X('playerID', title=''),
    y = alt.Y('RF', title='Range Factor')
)

line2 = alt.Chart(dffactorsuccess).mark_rule(color='green').encode(
    y = alt.Y('RF', title='Range Factor'),
)

graph3 = alt.Chart(dffactor).mark_circle(color='blue').encode(
    x = alt.X('playerID', title=''),
    y = alt.Y('RF', title='Range Factor'),
    size = 'G'
)

line3 = alt.Chart(dffactor).mark_rule(color='blue').encode(
    y = alt.Y('mean(RF)', title='Range Factor'),
)

(graph1 + graph2 + graph3 + line1 + line2 + line3)

alt.LayerChart(...)

**Insights:**
<br>1. Albert's career factor range (green) nearly double that of the MLB average (3.083).
<br>2. Clint's career factor range (red), on the other hand), is under the MLB averge. This shows us a bit about why he is considered a first draft botch. The dude didn't do so hot when it came to defense.

## K-Nearest Neighbors

### Machine Learning Libraries

In [94]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score
from scipy.optimize import curve_fit
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import f1_score
from sklearn.cluster import KMeans
from sklearn.feature_selection import mutual_info_classif

In [99]:
# Following code is providded courtesy of Prof. Danielle Szafir, INFO 3401.

def runKNN(df, x, y, k): 
    # Let's assume that we're using numeric features to predict a categorical label
    X = x.values
    Y = y.values.reshape(-1, 1)
    X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.2, random_state=1, stratify=Y)
    
    # Build a kNN classifier
    knn = KNeighborsClassifier(n_neighbors=k)
    knn.fit(X_train, Y_train)
    
    # Compute the quality of our predictions
    score = knn.score(X_test, Y_test)
    print("Predicts " + y.name + " with " + str(score) + " accuracy")
    print("F1 score is " + str(f1_score(Y_test, knn.predict(X_test), labels=y.unique(), average='weighted')))
    print("Chance is: " + str(1.0 / len(y.unique())))
    
    return knn

In [152]:
kdffielding = dfmasterfiel.fillna(0)

In [153]:
kdffielding.drop(columns=['yearID', 'stint', 'teamID', 'lgID', ], inplace=True)

In [154]:
kdffielding

playerID POS   G   GS  InnOuts    PO     A     E   DP   PB   WP   SB  \
0       abercda01  SS   1  0.0      0.0   1.0   3.0   2.0  0.0  0.0  0.0  0.0   
1        addybo01  2B  22  0.0      0.0  67.0  72.0  42.0  5.0  0.0  0.0  0.0   
2        addybo01  SS   3  0.0      0.0   8.0  14.0   7.0  0.0  0.0  0.0  0.0   
3       allisar01  2B   2  0.0      0.0   1.0   4.0   0.0  0.0  0.0  0.0  0.0   
4       allisar01  OF  29  0.0      0.0  51.0   3.0   7.0  1.0  0.0  0.0  0.0   
...           ...  ..  ..  ...      ...   ...   ...   ...  ...  ...  ...  ...   
170521  younger03  OF  22  0.0      0.0  45.0   0.0   0.0  0.0  0.0  0.0  0.0   
170522  younger03  OF   7  0.0      0.0   5.0   0.0   0.0  0.0  0.0  0.0  0.0   
170523  zimmery01  OF   1  0.0      0.0   0.0   0.0   0.0  0.0  0.0  0.0  0.0   
170524  zobribe01  OF  29  0.0      0.0  39.0   0.0   3.0  0.0  0.0  0.0  0.0   
170525  zobribe01  OF  20  0.0      0.0  36.0   1.0   1.0  0.0  0.0  0.0  0.0   

         CS   ZR  
0       0.0  0.0  
1       0.0  0.0  
2       0.0  0.0  
3       0.0  0.0  
4       0.0  0.0  
...     ...  ...  
170521  0.0  0.0  
170522  0.0  0.0  
170523  0.0  0.0  
170524  0.0  0.0  
170525  0.0  0.0  

[170526 rows x 14 columns]

In [159]:
aggregation_functions = {'POS': 'first', 'G': 'sum', 'GS': 'sum', 'InnOuts': 'sum', 'PO': 'sum', 
                         'A': 'sum', 'E': 'sum', 'DP': 'sum', 'PB': 'sum', 'WP': 'sum', 'SB': 'sum', 'CS': 'sum', 'ZR': 'sum'}
kdffielding = kdffielding.groupby(kdffielding['playerID']).aggregate(aggregation_functions)

In [165]:
kdffielding.reset_index(inplace=True)

In [166]:
kdffielding

playerID POS     G      GS   InnOuts       PO      A      E     DP  \
0      aardsda01   P   331     0.0     921.0     11.0   29.0    3.0    2.0   
1      aaronha01  LF  6019  5712.0  150488.0  12926.0  647.0  259.0  258.0   
2      aaronto01  1B   525   317.0    9346.0   1494.0  117.0   32.0  124.0   
3       aasedo01   P   448    91.0    3328.0     67.0  135.0   13.0   10.0   
4       abadan01  1B    10     0.0     138.0     37.0    1.0    1.0    3.0   
...          ...  ..   ...     ...       ...      ...    ...    ...    ...   
18460   zupofr01   C    13     1.0     114.0     31.0    1.0    2.0    0.0   
18461  zuvelpa01  SS   204   136.0    3844.0    267.0  415.0   23.0   84.0   
18462  zuverge01   P   265    31.0    1847.0     45.0  145.0    7.0   10.0   
18463  zwilldu01  OF   342     0.0       0.0    755.0   37.0   25.0   10.0   
18464   zychto01   P    13     0.0       0.0      0.0    3.0    0.0    0.0   

        PB   WP   SB   CS   ZR  
0      0.0  0.0  0.0  0.0  0.0  
1      0.0  0.0  0.0  0.0  0.0  
2      0.0  0.0  0.0  0.0  0.0  
3      0.0  0.0  0.0  0.0  0.0  
4      0.0  0.0  0.0  0.0  0.0  
...    ...  ...  ...  ...  ...  
18460  1.0  1.0  2.0  1.0  0.0  
18461  0.0  0.0  0.0  0.0  0.0  
18462  0.0  0.0  0.0  0.0  0.0  
18463  0.0  0.0  0.0  0.0  0.0  
18464  0.0  0.0  0.0  0.0  0.0  

[18465 rows x 14 columns]

## Train a k-NN Classifier

**Predict the position of the player as a function of errors.**

Errors are a good indicator because you will typically see less errors in in-field positions versus out-field positions, and almost never see errors from pitchers and catchers given the role that those positions play.

In [167]:
x = kdffielding[['E']]
y = kdffielding['POS']

In [113]:
# find the best k to use

for k in range(1, 10):
    print(str(k), 'neighbors: ')
    runKNN(kdffielding, x, y, k)

1 neighbors: 


C:\Users\jkibo\anaconda3\lib\site-packages\ipykernel_launcher.py:11: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  # This is added back by InteractiveShellApp.init_path()


Predicts POS with 0.08980824488359819 accuracy
F1 score is 0.060987453650665685
Chance is: 0.09090909090909091
2 neighbors: 


C:\Users\jkibo\anaconda3\lib\site-packages\ipykernel_launcher.py:11: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  # This is added back by InteractiveShellApp.init_path()


Predicts POS with 0.08866475107019293 accuracy
F1 score is 0.051815646785461154
Chance is: 0.09090909090909091
3 neighbors: 


C:\Users\jkibo\anaconda3\lib\site-packages\ipykernel_launcher.py:11: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  # This is added back by InteractiveShellApp.init_path()


Predicts POS with 0.08972028382102856 accuracy
F1 score is 0.05521436150695697
Chance is: 0.09090909090909091
4 neighbors: 


C:\Users\jkibo\anaconda3\lib\site-packages\ipykernel_launcher.py:11: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  # This is added back by InteractiveShellApp.init_path()


Predicts POS with 0.08584999706796458 accuracy
F1 score is 0.050348408616358045
Chance is: 0.09090909090909091
5 neighbors: 


C:\Users\jkibo\anaconda3\lib\site-packages\ipykernel_launcher.py:11: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  # This is added back by InteractiveShellApp.init_path()


Predicts POS with 0.10622764322993022 accuracy
F1 score is 0.062304751650215956
Chance is: 0.09090909090909091
6 neighbors: 


C:\Users\jkibo\anaconda3\lib\site-packages\ipykernel_launcher.py:11: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  # This is added back by InteractiveShellApp.init_path()


Predicts POS with 0.10508414941652496 accuracy
F1 score is 0.061919554541159126
Chance is: 0.09090909090909091
7 neighbors: 


C:\Users\jkibo\anaconda3\lib\site-packages\ipykernel_launcher.py:11: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  # This is added back by InteractiveShellApp.init_path()


Predicts POS with 0.12769014249692137 accuracy
F1 score is 0.08864902145294688
Chance is: 0.09090909090909091
8 neighbors: 


C:\Users\jkibo\anaconda3\lib\site-packages\ipykernel_launcher.py:11: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  # This is added back by InteractiveShellApp.init_path()


Predicts POS with 0.12930276197736468 accuracy
F1 score is 0.08819493216709834
Chance is: 0.09090909090909091
9 neighbors: 


C:\Users\jkibo\anaconda3\lib\site-packages\ipykernel_launcher.py:11: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  # This is added back by InteractiveShellApp.init_path()


Predicts POS with 0.1250219902656424 accuracy
F1 score is 0.1170074883847625
Chance is: 0.09090909090909091


8 neighbors is the best k value, with an accuracy of 0.129

In [168]:
def findName(playerID):
    first = dfmaster[dfmaster['playerID'] == playerID]['nameFirst'].reset_index()['nameFirst'][0]
    last = dfmaster[dfmaster['playerID'] == playerID]['nameLast'].reset_index()['nameLast'][0]
    return str(first)+' '+str(last)

analyze player performance using model:

In [169]:
# Clint Hurdle: hurdlcl01 (botch)
player1 = kdffielding[kdffielding['playerID'] == 'hurdlcl01']
player1Data = player1.select_dtypes(include='number')

# Albert Pujols: pujolal01 (success)
player2 = kdffielding[kdffielding['playerID'] == 'pujolal01']
player2Data = player2.select_dtypes(include='number')

In [170]:
# build prediction models
x = kdffielding.select_dtypes(include="number")

model = runKNN(kdffielding, x, y, 8)

print("k-NN thinks", player1["playerID"], "is a", model.predict(player1Data)[0])
print("k-NN thinks", player2["playerID"], "is a", model.predict(player2Data)[0])

C:\Users\jkibo\anaconda3\lib\site-packages\ipykernel_launcher.py:11: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  # This is added back by InteractiveShellApp.init_path()


Predicts POS with 0.7221770917952883 accuracy
F1 score is 0.6973440634557723
Chance is: 0.09090909090909091
k-NN thinks 7933    hurdlcl01
Name: playerID, dtype: object is a CF
k-NN thinks 13454    pujolal01
Name: playerID, dtype: object is a 1B


C:\Users\jkibo\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1515: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 in labels with no true nor predicted samples. Use `zero_division` parameter to control this behavior.
  average, "true nor predicted", 'F-score is', len(true_sum)


In [171]:
relatedPlayers1 = model.kneighbors(player1Data, n_neighbors=3, return_distance=False)
relatedPlayers2 = model.kneighbors(player2Data, n_neighbors=3, return_distance=False)

### K-Nearest Neighbors: Clint Hurdle

In [174]:
for p in relatedPlayers1[0]:
    playprint = kdffielding.iloc[int(p)]["playerID"]
    print(playprint, ':', findName(playprint))

repulri01 : Rip Repulski
garrewa01 : Wayne Garrett
howarmi01 : Mike Howard


In [180]:
# Clint Hurdle
dfmasterfiel[dfmasterfiel['playerID'] == 'hurdlcl01']

playerID  yearID  stint teamID lgID POS    G     GS  InnOuts     PO  \
84482   hurdlcl01    1977      1    KCA   AL  OF    9    6.0    185.0   17.0   
84483   hurdlcl01    1977      1    KCA   AL  RF    9    6.0    185.0   16.0   
86409   hurdlcl01    1978      1    KCA   AL  1B   52   49.0   1189.0  412.0   
86410   hurdlcl01    1978      1    KCA   AL  3B    1    0.0      3.0    0.0   
86411   hurdlcl01    1978      1    KCA   AL  DH    1    NaN      NaN    NaN   
86412   hurdlcl01    1978      1    KCA   AL  LF   41   35.0    843.0   59.0   
86413   hurdlcl01    1978      1    KCA   AL  OF   78   68.0   1753.0  132.0   
86414   hurdlcl01    1978      1    KCA   AL  RF   42   33.0    910.0   75.0   
88258   hurdlcl01    1979      1    KCA   AL  3B    1    0.0      3.0    1.0   
88259   hurdlcl01    1979      1    KCA   AL  DH    4    NaN      NaN    NaN   
88260   hurdlcl01    1979      1    KCA   AL  LF   30   28.0    606.0   40.0   
88261   hurdlcl01    1979      1    KCA   AL  OF   50   47.0   1113.0   88.0   
88262   hurdlcl01    1979      1    KCA   AL  RF   20   19.0    507.0   47.0   
90183   hurdlcl01    1980      1    KCA   AL  LF    1    0.0     15.0    0.0   
90184   hurdlcl01    1980      1    KCA   AL  OF  126  109.0   2931.0  233.0   
90185   hurdlcl01    1980      1    KCA   AL  RF  125  109.0   2916.0  236.0   
92080   hurdlcl01    1981      1    KCA   AL  OF   28   26.0    600.0   59.0   
92081   hurdlcl01    1981      1    KCA   AL  RF   28   26.0    600.0   59.0   
93909   hurdlcl01    1982      1    CIN   NL  LF   17    9.0    255.0   16.0   
93910   hurdlcl01    1982      1    CIN   NL  OF   17    9.0    258.0   17.0   
93911   hurdlcl01    1982      1    CIN   NL  RF    1    0.0      3.0    0.0   
95883   hurdlcl01    1983      1    NYN   NL  3B    9    8.0    174.0    1.0   
95884   hurdlcl01    1983      1    NYN   NL  OF    1    1.0     27.0    0.0   
95885   hurdlcl01    1983      1    NYN   NL  RF    1    1.0     27.0    0.0   
99877   hurdlcl01    1985      1    NYN   NL   C   17   12.0    344.0   79.0   
99878   hurdlcl01    1985      1    NYN   NL  LF    2    1.0     54.0    1.0   
99879   hurdlcl01    1985      1    NYN   NL  OF   10    7.0    222.0   10.0   
99880   hurdlcl01    1985      1    NYN   NL  RF    8    6.0    168.0    9.0   
101815  hurdlcl01    1986      1    SLN   NL  1B   39   30.0    814.0  301.0   
101816  hurdlcl01    1986      1    SLN   NL  3B    4    1.0     53.0    2.0   
101817  hurdlcl01    1986      1    SLN   NL   C    5    3.0     77.0   19.0   
101818  hurdlcl01    1986      1    SLN   NL  LF    5    2.0     69.0    3.0   
101819  hurdlcl01    1986      1    SLN   NL  OF   10    6.0    156.0   12.0   
101820  hurdlcl01    1986      1    SLN   NL  RF    5    4.0     87.0    9.0   
103874  hurdlcl01    1987      1    NYN   NL  1B    1    0.0      6.0    1.0   

           A     E    DP   PB   WP    SB   CS   ZR  
84482    0.0   0.0   0.0  NaN  NaN   NaN  NaN  NaN  
84483    0.0   0.0   0.0  NaN  NaN   NaN  NaN  NaN  
86409   25.0   6.0  47.0  NaN  NaN   NaN  NaN  NaN  
86410    0.0   0.0   0.0  NaN  NaN   NaN  NaN  NaN  
86411    NaN   NaN   NaN  NaN  NaN   NaN  NaN  NaN  
86412    2.0   2.0   1.0  NaN  NaN   NaN  NaN  NaN  
86413    5.0   6.0   1.0  NaN  NaN   NaN  NaN  NaN  
86414    3.0   4.0   0.0  NaN  NaN   NaN  NaN  NaN  
88258    0.0   0.0   0.0  NaN  NaN   NaN  NaN  NaN  
88259    NaN   NaN   NaN  NaN  NaN   NaN  NaN  NaN  
88260    3.0   1.0   0.0  NaN  NaN   NaN  NaN  NaN  
88261    2.0   3.0   0.0  NaN  NaN   NaN  NaN  NaN  
88262    0.0   2.0   0.0  NaN  NaN   NaN  NaN  NaN  
90183    0.0   0.0   0.0  NaN  NaN   NaN  NaN  NaN  
90184    8.0  10.0   1.0  NaN  NaN   NaN  NaN  NaN  
90185    8.0   9.0   1.0  NaN  NaN   NaN  NaN  NaN  
92080    1.0   0.0   0.0  NaN  NaN   NaN  NaN  NaN  
92081    1.0   0.0   0.0  NaN  NaN   NaN  NaN  NaN  
93909    2.0   1.0   0.0  NaN  NaN   NaN  NaN  NaN  
93910    2.0   1.0   0.0  NaN  NaN   NaN  NaN  NaN  
93911    0.0   

In [176]:
# Rip Repulski
dfmasterfiel[dfmasterfiel['playerID'] == 'repulri01']

playerID  yearID  stint teamID lgID POS    G     GS  InnOuts     PO  \
51268  repulri01    1953      1    SLN   NL  OF  153    NaN      NaN  361.0   
52137  repulri01    1954      1    SLN   NL  CF   14   13.0    354.0   38.0   
52138  repulri01    1954      1    SLN   NL  LF  137  133.0   3666.0  259.0   
52139  repulri01    1954      1    SLN   NL  OF  152  149.0   4093.0  302.0   
52140  repulri01    1954      1    SLN   NL  RF    3    3.0     73.0    3.0   
53189  repulri01    1955      1    SLN   NL  CF    1    0.0     21.0    2.0   
53190  repulri01    1955      1    SLN   NL  LF  115  109.0   2974.0  211.0   
53191  repulri01    1955      1    SLN   NL  OF  141  131.0   3595.0  260.0   
53192  repulri01    1955      1    SLN   NL  RF   27   22.0    600.0   43.0   
54263  repulri01    1956      1    SLN   NL  CF    1    0.0      3.0    0.0   
54264  repulri01    1956      1    SLN   NL  LF  100   93.0   2485.0  183.0   
54265  repulri01    1956      1    SLN   NL  OF  100   93.0   2494.0  187.0   
54266  repulri01    1956      1    SLN   NL  RF    1    0.0      6.0    0.0   
55306  repulri01    1957      1    PHI   NL  LF   52   41.0   1126.0   84.0   
55307  repulri01    1957      1    PHI   NL  OF  130  129.0   3450.0  264.0   
55308  repulri01    1957      1    PHI   NL  RF   91   88.0   2324.0  179.0   
56407  repulri01    1958      1    PHI   NL  LF   36   35.0    903.0   58.0   
56408  repulri01    1958      1    PHI   NL  OF   56   53.0   1365.0   90.0   
56409  repulri01    1958      1    PHI   NL  RF   20   18.0    462.0   34.0   
57512  repulri01    1959      1    LAN   NL  CF    2    2.0     42.0    1.0   
57513  repulri01    1959      1    LAN   NL  LF   16   11.0    267.0   17.0   
57514  repulri01    1959      1    LAN   NL  OF   31   20.0    512.0   30.0   
57515  repulri01    1959      1    LAN   NL  RF   13    7.0    203.0   12.0   
58610  repulri01    1960      2    BOS   AL  LF   33   22.0    663.0   53.0   
58611  repulri01    1960      1    LAN   NL  OF    2    1.0     30.0    1.0   
58612  repulri01    1960      2    BOS   AL  OF   33   22.0    663.0   56.0   
58613  repulri01    1960      1    LAN   NL  RF    2    1.0     30.0    1.0   
59785  repulri01    1961      1    BOS   AL  LF    4    4.0     87.0    4.0   
59786  repulri01    1961      1    BOS   AL  OF    4    4.0     87.0    4.0   

         A    E   DP  PB  WP  SB  CS  ZR  
51268  7.0  5.0  1.0 NaN NaN NaN NaN NaN  
52137  0.0  1.0  0.0 NaN NaN NaN NaN NaN  
52138  5.0  7.0  0.0 NaN NaN NaN NaN NaN  
52139  4.0  8.0  0.0 NaN NaN NaN NaN NaN  
52140  0.0  1.0  0.0 NaN NaN NaN NaN NaN  
53189  0.0  0.0  0.0 NaN NaN NaN NaN NaN  
53190  2.0  6.0  1.0 NaN NaN NaN NaN NaN  
53191  5.0  7.0  1.0 NaN NaN NaN NaN NaN  
53192  2.0  1.0  0.0 NaN NaN NaN NaN NaN  
54263  0.0  0.0  0.0 NaN NaN NaN NaN NaN  
54264  2.0  5.0  0.0 NaN NaN NaN NaN NaN  
54265  3.0  5.0  0.0 NaN NaN NaN NaN NaN  
54266  0.0  0.0  0.0 NaN NaN NaN NaN NaN  
55306  1.0  2.0  1.0 NaN NaN NaN NaN NaN  
55307  6.0  9.0  2.0 NaN NaN NaN NaN NaN  
55308  4.0  8.0  1.0 NaN NaN NaN NaN NaN  
56407  0.0  3.0  0.0 NaN NaN NaN NaN NaN  
56408  3.0  5.0  0.0 NaN NaN NaN NaN NaN  
56409  3.0  2.0  0.0 NaN NaN NaN NaN NaN  
57512  0.0  0.0  0.0 NaN NaN NaN NaN NaN  
57513  0.0  0.0  0.0 NaN NaN NaN NaN NaN  
57514  0.0  0.0  0.0 NaN NaN NaN NaN NaN  
57515  0.0  0.0  0.0 NaN NaN NaN NaN NaN  
58610  0.0  0.0  0.0 NaN NaN NaN NaN NaN  
58611  0.0  0.0  0.0 NaN NaN NaN NaN NaN  
58612  0.0  0.0  0.0 NaN NaN NaN NaN NaN  
58613  0.0  0.0  0.0 NaN NaN NaN NaN NaN  
59785  0.0  0.0  0.0 NaN NaN NaN NaN NaN  
59786  0.0  0.0  0.0 NaN NaN NaN NaN NaN

In [178]:
# Wayne Garrett
dfmasterfiel[dfmasterfiel['playerID'] == 'garrewa01']

playerID  yearID  stint teamID lgID POS    G     GS  InnOuts     PO  \
70113  garrewa01    1969      1    NYN   NL  2B   47   34.0    988.0  102.0   
70114  garrewa01    1969      1    NYN   NL  3B   72   63.0   1752.0   40.0   
70115  garrewa01    1969      1    NYN   NL  SS    9    1.0     72.0    5.0   
71875  garrewa01    1970      1    NYN   NL  2B   45   34.0   1001.0   94.0   
71876  garrewa01    1970      1    NYN   NL  3B   70   67.0   1815.0   57.0   
71877  garrewa01    1970      1    NYN   NL  SS    1    0.0     15.0    1.0   
73620  garrewa01    1971      1    NYN   NL  2B    9    6.0    174.0   18.0   
73621  garrewa01    1971      1    NYN   NL  3B   53   45.0   1259.0   30.0   
75288  garrewa01    1972      1    NYN   NL  2B   22   18.0    472.0   48.0   
75289  garrewa01    1972      1    NYN   NL  3B   82   70.0   1937.0   66.0   
77017  garrewa01    1973      1    NYN   NL  2B    6    6.0    162.0   14.0   
77018  garrewa01    1973      1    NYN   NL  3B  130  118.0   3213.0   80.0   
77019  garrewa01    1973      1    NYN   NL  SS    9    6.0    190.0    8.0   
78765  garrewa01    1974      1    NYN   NL  3B  144  136.0   3713.0  111.0   
78766  garrewa01    1974      1    NYN   NL  SS    9    6.0    195.0   12.0   
80557  garrewa01    1975      1    NYN   NL  2B    1    0.0      3.0    0.0   
80558  garrewa01    1975      1    NYN   NL  3B   94   68.0   1952.0   64.0   
80559  garrewa01    1975      1    NYN   NL  SS    3    0.0     13.0    1.0   
82360  garrewa01    1976      1    NYN   NL  2B   10    9.0    258.0   25.0   
82361  garrewa01    1976      2    MON   NL  2B   54   44.0   1271.0  121.0   
82362  garrewa01    1976      1    NYN   NL  3B   64   58.0   1588.0   47.0   
82363  garrewa01    1976      2    MON   NL  3B    2    1.0     24.0    1.0   
82364  garrewa01    1976      1    NYN   NL  SS    1    0.0      3.0    1.0   
84269  garrewa01    1977      1    MON   NL  2B    1    0.0      3.0    0.0   
84270  garrewa01    1977      1    MON   NL  3B   49   44.0   1197.0   34.0   
86212  garrewa01    1978      1    MON   NL  3B   13   11.0    303.0    6.0   
86213  garrewa01    1978      2    SLN   NL  3B   19   17.0    384.0   12.0   

           A     E    DP  PB  WP  SB  CS  ZR  
70113   97.0   3.0  27.0 NaN NaN NaN NaN NaN  
70114  115.0   8.0  10.0 NaN NaN NaN NaN NaN  
70115    6.0   0.0   0.0 NaN NaN NaN NaN NaN  
71875   92.0   2.0  28.0 NaN NaN NaN NaN NaN  
71876  113.0  10.0   6.0 NaN NaN NaN NaN NaN  
71877    0.0   0.0   0.0 NaN NaN NaN NaN NaN  
73620   11.0   0.0   4.0 NaN NaN NaN NaN NaN  
73621   89.0   4.0   7.0 NaN NaN NaN NaN NaN  
75288   38.0   5.0   6.0 NaN NaN NaN NaN NaN  
75289  150.0   9.0  13.0 NaN NaN NaN NaN NaN  
77017   18.0   1.0   6.0 NaN NaN NaN NaN NaN  
77018  280.0  22.0  36.0 NaN NaN NaN NaN NaN  
77019   15.0   3.0   4.0 NaN NaN NaN NaN NaN  
78765  318.0  20.0  31.0 NaN NaN NaN NaN NaN  
78766   31.0   0.0   2.0 NaN NaN NaN NaN NaN  
80557    1.0   0.0   0.0 NaN NaN NaN NaN NaN  
80558  160.0   8.0  24.0 NaN NaN NaN NaN NaN  
80559    1.0   0.0   0.0 NaN NaN NaN NaN NaN  
82360   18.0   0.0   3.0 NaN NaN NaN NaN NaN  
82361  157.0   5.0  32.0 NaN NaN NaN NaN NaN  
82362  137.0  10.0  12.0 NaN NaN NaN NaN NaN  
82363    1.0   0.0   0.0 NaN NaN NaN NaN NaN  
82364    1.0   0.0   1.0 NaN NaN NaN NaN NaN  
84269    0.0   0.0   0.0 NaN NaN NaN NaN NaN  
84270  101.0   0.0   2.0 NaN NaN NaN NaN NaN  
86212   25.0   1.0   2.0 NaN NaN NaN NaN NaN  
86213   26.0   3.0   5.0 NaN NaN NaN NaN NaN

In [179]:
# Mike Howard
dfmasterfiel[dfmasterfiel['playerID'] == 'howarmi01']

playerID  yearID  stint teamID lgID POS   G   GS  InnOuts    PO    A  \
92061  howarmi01    1981      1    NYN   NL  CF   1  1.0     27.0   0.0  1.0   
92062  howarmi01    1981      1    NYN   NL  LF   6  0.0     30.0   4.0  0.0   
92063  howarmi01    1981      1    NYN   NL  OF  14  6.0    208.0  18.0  2.0   
92064  howarmi01    1981      1    NYN   NL  RF   7  5.0    151.0  13.0  1.0   
93890  howarmi01    1982      1    NYN   NL  2B   3  0.0     36.0   5.0  4.0   
93891  howarmi01    1982      1    NYN   NL  CF   7  6.0    156.0  22.0  1.0   
93892  howarmi01    1982      1    NYN   NL  LF   9  0.0     58.0   5.0  1.0   
93893  howarmi01    1982      1    NYN   NL  OF  22  7.0    283.0  27.0  2.0   
93894  howarmi01    1982      1    NYN   NL  RF   8  1.0     69.0   0.0  0.0   
95866  howarmi01    1983      1    NYN   NL  OF   1  1.0     27.0   0.0  0.0   
95867  howarmi01    1983      1    NYN   NL  RF   1  1.0     27.0   0.0  0.0   

         E   DP  PB  WP  SB  CS  ZR  
92061  0.0  0.0 NaN NaN NaN NaN NaN  
92062  0.0  0.0 NaN NaN NaN NaN NaN  
92063  1.0  0.0 NaN NaN NaN NaN NaN  
92064  1.0  0.0 NaN NaN NaN NaN NaN  
93890  0.0  1.0 NaN NaN NaN NaN NaN  
93891  0.0  0.0 NaN NaN NaN NaN NaN  
93892  0.0  0.0 NaN NaN NaN NaN NaN  
93893  0.0  0.0 NaN NaN NaN NaN NaN  
93894  0.0  0.0 NaN NaN NaN NaN NaN  
95866  0.0  0.0 NaN NaN NaN NaN NaN  
95867  0.0  0.0 NaN NaN NaN NaN NaN

### K-Nearest Neighbors: Albert Pujols

In [175]:
for p in relatedPlayers2[0]:
    playprint = kdffielding.iloc[int(p)]["playerID"]
    print(playprint, ':', findName(playprint))

mateoju02 : Juan Mateo
baumeji01 : Jim Baumer
derricl01 : Claud Derrick


In [181]:
# Albert Pujols
dfmasterfiel[dfmasterfiel['playerID'] == 'pujolal01']

playerID  yearID  stint teamID lgID POS    G  GS  InnOuts      PO  \
136585  pujolal01    2001      1    SLN   NL  1B   42 NaN    861.0   283.0   
136586  pujolal01    2001      1    SLN   NL  3B   55 NaN   1296.0    40.0   
136587  pujolal01    2001      1    SLN   NL  LF   39 NaN      NaN     NaN   
136588  pujolal01    2001      1    SLN   NL  OF   78 NaN   1836.0   128.0   
136589  pujolal01    2001      1    SLN   NL  RF   39 NaN      NaN     NaN   
138936  pujolal01    2002      1    SLN   NL  1B   21 NaN    432.0   140.0   
138937  pujolal01    2002      1    SLN   NL  3B   41 NaN    879.0    25.0   
138938  pujolal01    2002      1    SLN   NL  LF  117 NaN      NaN     NaN   
138939  pujolal01    2002      1    SLN   NL  OF  118 NaN   2622.0   173.0   
138940  pujolal01    2002      1    SLN   NL  RF    1 NaN      NaN     NaN   
138941  pujolal01    2002      1    SLN   NL  SS    1 NaN      6.0     0.0   
141227  pujolal01    2003      1    SLN   NL  1B   62 NaN   1110.0   340.0   
141228  pujolal01    2003      1    SLN   NL  LF  113 NaN      NaN     NaN   
141229  pujolal01    2003      1    SLN   NL  OF  113 NaN   2712.0   198.0   
143622  pujolal01    2004      1    SLN   NL  1B  150 NaN   4017.0  1458.0   
145851  pujolal01    2005      1    SLN   NL  1B  157 NaN   4077.0  1597.0   
148190  pujolal01    2006      1    SLN   NL  1B  143 NaN   3732.0  1348.0   
150594  pujolal01    2007      1    SLN   NL  1B  154 NaN   3975.0  1325.0   
152946  pujolal01    2008      1    SLN   NL  1B  144 NaN   3645.0  1297.0   
152947  pujolal01    2008      1    SLN   NL  2B    1 NaN      9.0     1.0   
155255  pujolal01    2009      1    SLN   NL  1B  159 NaN   4131.0  1473.0   
157580  pujolal01    2010      1    SLN   NL  1B  157 NaN   4143.0  1458.0   
159928  pujolal01    2011      1    SLN   NL  1B  146 NaN   3780.0  1254.0   
159929  pujolal01    2011      1    SLN   NL  3B    7 NaN    102.0     2.0   
162417  pujolal01    2012      1    LAA   AL  1B  120 NaN   3066.0  1013.0   
162418  pujolal01    2012      1    LAA   AL  3B    3 NaN     48.0     1.0   
164852  pujolal01    2013      1    LAA   AL  1B   34 NaN    906.0   265.0   
167275  pujolal01    2014      1    LAA   AL  1B  116 NaN   3051.0   879.0   
167276  pujolal01    2014      1    LAA   AL  3B    1 NaN     18.0     0.0   
169009  pujolal01    2015      1    LAA   AL  1B   95 NaN      NaN   718.0   
169347  pujolal01    2015      1    LAA   AL  3B    1 NaN      NaN     0.0   

            A     E     DP  PB  WP  SB  CS  ZR  
136585   19.0   5.0   27.0 NaN NaN NaN NaN NaN  
136586  111.0  10.0   17.0 NaN NaN NaN NaN NaN  
136587    NaN   NaN    NaN NaN NaN NaN NaN NaN  
136588    6.0   5.0    0.0 NaN NaN NaN NaN NaN  
136589    NaN   NaN    NaN NaN NaN NaN NaN NaN  
138936   13.0   1.0   24.0 NaN NaN NaN NaN NaN  
138937   66.0   6.0    6.0 NaN NaN NaN NaN NaN  
138938    NaN   NaN    NaN NaN NaN NaN NaN NaN  
138939    4.0   4.0    0.0 NaN NaN NaN NaN NaN  
138940    NaN   NaN    NaN NaN NaN NaN NaN NaN  
138941    0.0   0.0    0.0 NaN NaN NaN NaN NaN  
141227   33.0   1.0   34.0 NaN NaN NaN NaN NaN  
141228    NaN   NaN    NaN NaN NaN NaN NaN NaN  
141229    7.0   3.0    0.0 NaN NaN NaN NaN NaN  
143622  114.0  10.0  136.0 NaN NaN NaN NaN NaN  
145851   97.0  14.0  175.0 NaN NaN NaN NaN NaN  
148190  110.0   6.0  145.0 NaN NaN NaN NaN NaN  
150594  124.0   8.0  132.0 NaN NaN NaN NaN NaN  
152946  135.0   6.0  119.0 NaN NaN NaN NaN NaN  
152947    0.0   0.0    0.0 NaN NaN NaN NaN NaN  
155255  185.0  13.0  150.0 NaN NaN NaN NaN NaN  
157580  157.0   4.0  146.0 NaN NaN NaN NaN NaN  
159928  165.0  11.0  134.0 NaN NaN NaN NaN NaN  
159929   12.0   3.0    2.0 NaN NaN NaN NaN NaN  
162417   97.0   7.0   95.0 NaN NaN NaN NaN NaN  
162418    3.0   0.0    2.0 NaN NaN NaN NaN NaN  
164852   29.0   3.0   24.0 NaN NaN NaN NaN NaN  
167275   80.0   3.0   78.0 NaN NaN NaN NaN NaN  
167276    1.0   0.0    0.0 NaN NaN NaN NaN NaN  
169009   55.0   5.0   57.0 NaN NaN NaN NaN

In [183]:
# Juan Mateo
dfmasterfiel[dfmasterfiel['playerID'] == 'mateoju02']

playerID  yearID  stint teamID lgID POS   G  GS  InnOuts   PO    A  \
147836  mateoju02    2006      1    CHN   NL   P  11 NaN    138.0  0.0  5.0   

          E   DP  PB  WP  SB  CS  ZR  
147836  0.0  1.0 NaN NaN NaN NaN NaN

In [184]:
# Jim Baumer
dfmasterfiel[dfmasterfiel['playerID'] == 'baumeji01']

playerID  yearID  stint teamID lgID POS  G   GS  InnOuts    PO     A  \
47936  baumeji01    1949      1    CHA   AL  SS  7  NaN      NaN   3.0  12.0   
58962  baumeji01    1961      1    CIN   NL  2B  9  8.0    191.0  18.0  14.0   

         E   DP  PB  WP  SB  CS  ZR  
47936  1.0  4.0 NaN NaN NaN NaN NaN  
58962  0.0  2.0 NaN NaN NaN NaN NaN

In [185]:
# Claud Derrick
dfmasterfiel[dfmasterfiel['playerID'] == 'derricl01']

playerID  yearID  stint teamID lgID POS   G  GS  InnOuts    PO     A  \
20999  derricl01    1910      1    PHA   AL  SS   1 NaN      NaN   1.0   0.0   
21723  derricl01    1911      1    PHA   AL  1B   3 NaN      NaN   9.0   0.0   
21724  derricl01    1911      1    PHA   AL  2B  21 NaN      NaN  44.0  52.0   
21725  derricl01    1911      1    PHA   AL  3B   2 NaN      NaN   5.0   4.0   
21726  derricl01    1911      1    PHA   AL  SS   6 NaN      NaN  11.0  15.0   
22469  derricl01    1912      1    PHA   AL  SS  18 NaN      NaN  27.0  57.0   
23250  derricl01    1913      1    NYA   AL  2B   1 NaN      NaN   1.0   2.0   
23251  derricl01    1913      1    NYA   AL  3B   4 NaN      NaN   3.0   6.0   
23252  derricl01    1913      1    NYA   AL  SS  17 NaN      NaN  32.0  51.0   
24099  derricl01    1914      1    CIN   NL  SS   2 NaN      NaN   4.0   4.0   
24100  derricl01    1914      2    CHN   NL  SS  28 NaN      NaN  65.0  88.0   

          E   DP  PB  WP  SB  CS  ZR  
20999   1.0  0.0 NaN NaN NaN NaN NaN  
21723   0.0  0.0 NaN NaN NaN NaN NaN  
21724   4.0  7.0 NaN NaN NaN NaN NaN  
21725   0.0  0.0 NaN NaN NaN NaN NaN  
21726   0.0  0.0 NaN NaN NaN NaN NaN  
22469  11.0  5.0 NaN NaN NaN NaN NaN  
23250   0.0  2.0 NaN NaN NaN NaN NaN  
23251   2.0  0.0 NaN NaN NaN NaN NaN  
23252  12.0  7.0 NaN NaN NaN NaN NaN  
24099   1.0  1.0 NaN NaN NaN NaN NaN  
24100  18.0  9.0 NaN NaN NaN NaN NaN

## K-Means

In [186]:
# Following code is providded courtesy of Prof. Danielle Szafir, INFO 3401.

def runKMeans(df, k):
    # Prep the data by only choosing the numeric features in the data
    X = df.select_dtypes(include="number")
    
    # Run the k-Means algorithm
    kmeans = KMeans(n_clusters=k)

    # Train the model on the features we've selected
    kmeans.fit(X)
    
    # Add the data to our dataframe in a column called 'Cluster'. 
    # Cluster will give a numeric label corresponding to the index of the closest mean
    df["Cluster"] = pd.Series(kmeans.predict(X), index=df.index)
    
    # Return the model
    return kmeans

**Predict the position of the player as a function of errors.**

Errors are a good indicator because you will typically see less errors in in-field positions versus out-field positions, and almost never see errors from pitchers and catchers given the role that those positions play.

In [188]:
# find best k value

# PART 1
max_k = 30

scores = []
for k in range(1, max_k):
    model = runKMeans(kdffielding, k)
    scores.append(model.inertia_)
    print(k,'done!')

1 done!
2 done!
3 done!
4 done!
5 done!
6 done!
7 done!
8 done!
9 done!
10 done!
11 done!
12 done!
13 done!
14 done!
15 done!
16 done!
17 done!
18 done!
19 done!
20 done!
21 done!
22 done!
23 done!
24 done!
25 done!
26 done!
27 done!
28 done!
29 done!


In [190]:
# Part 2
scoresDf = pd.DataFrame()
scoresDf["k"] = range(1, max_k)
scoresDf["Score"] = scores

alt.Chart(scoresDf).mark_line().encode(
    x = "k",
    y = "Score"
)

alt.Chart(...)

using the graph above we can conclude the 3 is the best k to use

In [192]:
# create model for data
model = runKMeans(kdffielding, 3)

# Following code is providded courtesy of Prof. Danielle Szafir, INFO 3401.
features = kdffielding.select_dtypes(include="number")
for c in range(0, len(model.cluster_centers_)):
    print("***** Cluster " + str(c))
    for f in range(0, len(features.columns)):
        feature = features.columns[f]
        print("Mean " + feature + " = " + str(model.cluster_centers_[c][f]))

***** Cluster 0
Mean G = 182.63976225858096
Mean GS = 23.585676077286536
Mean InnOuts = 982.7229123338402
Mean PO = 354.5524517087922
Mean A = 168.85937592868677
Mean E = 23.13390787518709
Mean DP = 28.281664190195862
Mean PB = 2.648796433878286
Mean WP = 0.8622882615153438
Mean SB = 2.766775631499746
Mean CS = 1.2432095096593625
Mean ZR = 0.04713224368499849
Mean Cluster = -3.5110803153770576e-14
***** Cluster 1
Mean G = 2801.211678832098
Mean GS = 2129.963503649625
Mean InnOuts = 63490.75912408741
Mean PO = 6600.445255474476
Mean A = 1446.44525547446
Mean E = 149.8175182481743
Mean DP = 400.6253041362535
Mean PB = 4.708029197080284
Mean WP = 15.54744525547439
Mean SB = 40.84428223844284
Mean CS = 22.39172749391723
Mean ZR = 1.2992700729927062
Mean Cluster = 1.0000000000000002
***** Cluster 2
Mean G = 1232.1432058584107
Mean GS = 626.8771358828244
Mean InnOuts = 24273.585842147902
Mean PO = 3051.812855980496
Mean A = 1127.6777868185604
Mean E = 80.1570382424726
Mean DP = 286.851912123

In [210]:
alt.Chart(kdffielding[:5000]).mark_circle().encode(
    alt.X(alt.repeat("column"), type="quantitative"),
    alt.Y(alt.repeat("row"), type="quantitative"),
    color="Cluster:N",
    tooltip=["playerID"]
).properties(
    width=100,
    height=100
).repeat(
    row=["G", "A", "E", "DP", "PO"],
    column=["G", "A", "E", "DP", "PO"]
).interactive()

alt.RepeatChart(...)

In [199]:
import operator

In [200]:
# Following code is providded courtesy of Prof. Danielle Szafir, INFO 3401.

def computeInformationGain(x, y): 
    # Create a dictionary matching each column in x to a 
    # corresponding mutual information value
    feature_list = dict(zip(x.columns, mutual_info_classif(x, y, discrete_features=True)))
    
    # Sort the list from smallest to largest
    feature_list = sorted(feature_list.items(), key=operator.itemgetter(1))
    return feature_list

In [201]:
x = kdffielding.select_dtypes(include="number")
y = kdffielding['POS']

features = computeInformationGain(x, y)
print(features)

[('ZR', 0.04397939867236263), ('Cluster', 0.07036650355782793), ('WP', 0.07450337483720876), ('CS', 0.10173528240958679), ('SB', 0.10985040801561362), ('PB', 0.20269753306116434), ('E', 0.2371260638366367), ('GS', 0.3013847857129372), ('DP', 0.30362393366034957), ('A', 0.39096117077618786), ('G', 0.45355436209736), ('PO', 0.7274789111866553), ('InnOuts', 0.7776999205138351)]


In [203]:
top_features = features[4 : len(features)]
filteredDf = pd.DataFrame()
filteredDf["POS"] = kdffielding["POS"]
filteredDf["playerID"] = kdffielding["playerID"]
for feature in top_features: 
    filteredDf[feature[0]] = kdffielding[feature[0]]

In [205]:
x_filtered = filteredDf.select_dtypes(include="number")
knn_filtered = runKNN(filteredDf, x_filtered, y, 3)
knn = runKNN(kdffielding, x, y, 3)

C:\Users\jkibo\anaconda3\lib\site-packages\ipykernel_launcher.py:11: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  # This is added back by InteractiveShellApp.init_path()


Predicts POS with 0.7005144868670458 accuracy


C:\Users\jkibo\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1515: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 in labels with no true nor predicted samples. Use `zero_division` parameter to control this behavior.
  average, "true nor predicted", 'F-score is', len(true_sum)
C:\Users\jkibo\anaconda3\lib\site-packages\ipykernel_launcher.py:11: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  # This is added back by InteractiveShellApp.init_path()


F1 score is 0.6878438196943307
Chance is: 0.09090909090909091
Predicts POS with 0.7015976171134579 accuracy
F1 score is 0.6890966028296175
Chance is: 0.09090909090909091


C:\Users\jkibo\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1515: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 in labels with no true nor predicted samples. Use `zero_division` parameter to control this behavior.
  average, "true nor predicted", 'F-score is', len(true_sum)


In [209]:
alt.Chart(kdffielding[:5000]).mark_circle().encode(
    alt.X(alt.repeat("column"), type="quantitative"),
    alt.Y(alt.repeat("row"), type="quantitative"),
    color="Cluster:N",
    tooltip=["playerID"]
).properties(
    width=100,
    height=100
).repeat(
    row=["G", "A", "E", "DP", "PO"],
    column=["G", "A", "E", "DP", "PO"]
).interactive()

alt.RepeatChart(...)

## SYNTHESIS